---
title: quantification of single nuc-seq data (alsf-filbin) with salmon 
author: Stephanie Hicks
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(here)
```

List cells and tumor names 

```{r}
path_dir_fast <- "/fastscratch/myscratch/shicks1/alsf-filbin"
if(!dir.exists(file.path(path_dir_fast, "salmon_quants"))){
  dir.create(file.path(path_dir_fast, "salmon_quants"))
}

path_dir_home <- "/users/shicks1/data/alsf_filbin/sample"
tumor_names <- list.files(path_dir_home)

unique_cell_names <- NULL
if(!file.exists(file.path(path_dir_home, "unique_cell_names.txt"))){
  tumor_names <- tumor_names[!(tumor_names %in% c("Human_medians.tsv", 
                                                "Kriegstein_Human_cortex"))]
  for(tum in tumor_names){
    ids <- list.files(paste(path_dir_home, tum, sep="/"))
    ids <- unique(stringr::str_sub(ids, end=-11))
    unique_cell_names <- c(unique_cell_names, ids)
  }
  readr::write_lines(unique_cell_names,
                     path = file.path(path_dir_home, "unique_cell_names.txt"))
}

tumor_names <- tumor_names[!(tumor_names %in% c("Human_medians.tsv", 
                                                "Kriegstein_Human_cortex", 
                                                "unique_cell_names.txt"))]

```

# Quantification 

## Download files needed for Salmon

e.g. gene transfer format (GTF) is a file format used to hold information 
about gene structure. It is a tab-delimited text format based on
the general feature format (GFF), but contains some additional 
conventions specific to gene information.

```{bash}
cd /users/shicks1/data/reference/genomes/hsapiens/ENSEMBL/GRCh38/salmon

# download ensemble fastq cdna file 
wget ftp://ftp.ensembl.org/pub/release-98/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz

# download gtf file
wget ftp://ftp.ensembl.org/pub/release-98/gtf/homo_sapiens/Homo_sapiens.GRCh38.98.gtf.gz
```

Next, we need to create a set of pre-mRNA sequences to 
save to a FASTA file. This is needed because we have 
single nucleus RNA-seq data so we need to take into account 
the pre-mRNA and mRNA transcripts. This code kindly comes from 
[Charlotte Soneson](https://csoneson.github.io).

```{r}
suppressPackageStartupMessages({
  library(Biostrings)
  library(rtracklayer)
  library(GenomicFeatures)
  library(BSgenome)
})

## Get genomic sequence
genomefasta <- file.path("/users/shicks1/data/reference/genomes/hsapiens/ENSEMBL/GRCh38/salmon", 
                         "Homo_sapiens.GRCh38.cdna.all.fa.gz")

## Import FASTA file with Biostrings::readDNAStringSet
genome <- Biostrings::readDNAStringSet(genomefasta)
names(genome) <- sapply(strsplit(names(genome), " "), .subset, 1)

ingtf <- file.path("/users/shicks1/data/reference/genomes/hsapiens/ENSEMBL/GRCh38/salmon", 
                   "Homo_sapiens.GRCh38.98.gtf.gz")

## Read gtf file and group exons by transcript
txdb <- GenomicFeatures::makeTxDbFromGFF(ingtf)
grl <- GenomicFeatures::exonsBy(txdb, by = "tx", use.names = TRUE)

## Create the pre-mRNA coordinates by adding all introns to each transcript
grrange <- unlist(range(grl))

## Get the sequence of each pre-mRNA molecule
premrnaseq <- BSgenome::getSeq(x = genome, names = grrange)

## Save pre-mRNA sequences to fasta file
Biostrings::writeXStringSet(premrnaseq, filepath = outfasta)
```

Now we will put the pre-mRNA (`outfasta`) and 
mRNA (`Homo_sapiens.GRCh38.cdna.all.fa.gz`) FASTA files together. 

```{bash}

```


## Build salmon index 

```{bash}
cd /users/shicks1/data/reference/genomes/hsapiens/ENSEMBL/GRCh38/salmon

# create salmon index (3-5 mins)
salmon index -t Homo_sapiens.GRCh38.cdna.all.fa.gz -i ensemble.grch38_salmon-index-v0.14.1
```

## Running salmon 

See the `scripts/run-salmon.sh` file 

## Creating SummarizedExperiment object

```{r}
path_dir <- "/fastscratch/myscratch/shicks1/alsf-filbin"
tumor_names <- list.files(path_dir)
tumor_names <- tumor_names[!(tumor_names %in% 
                          c("Human_medians.tsv", "Kriegstein_Human_cortex", 
                            "unique_cell_names.txt", "salmon_quants"))]

unique_sf_paths <- NULL
for(tum in tumor_names){
  ids <- list.files(paste(path_dir, tum, sep="/"))
  ids <- unique(stringr::str_sub(ids, end=-13))
  ids <- paste(path_dir, "salmon_quants", paste0(ids, "_quant/quant.sf"), sep="/")
  unique_sf_paths <- c(unique_sf_paths, ids)
}

unique_sf_ids <- NULL
for(tum in tumor_names){
  ids <- list.files(paste(path_dir, tum, sep="/"))
  ids <- unique(stringr::str_sub(ids, end=-13))
  unique_sf_ids <- c(unique_sf_ids, ids)
}

coldata <- data.frame(files = unique_sf_paths, names = unique_sf_ids)

```

```{r}
library(tximeta)
se <- tximeta(coldata[1:96,])
```


# Quality control 

## Explore SummarizedExperiment object

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
colData(se)
assayNames(se)
rowRanges(se)
```

```{r}
dat <- assay(se, "counts")
hist(colSums(dat))
```