---
title: quantification of single nuc-seq data (alsf-filbin) with salmon 
author: Stephanie Hicks and Albert Kuo
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{bash}
# Commands to start an interactive session on the JHPCE cluster
qrsh -l mem_free=10G,h_vmem=10G
module load conda_R
cd /fastscratch/myscratch/akuo/alsf-filbin
R
```

```{r}
library(here)
```

This code chunk copies the data from where it is stored into the github repo so we can all work with relative paths. 
I copy it so I can keep the backed up copy where I originally have it. 
However, the person using this script needs to adjust which paths to use. 

```{r}
# stephanie_data_path <- "/users/shicks1/data/alsf_filbin/sample"
albert_data_path <- "/users/akuo/alsf-filbin/sample_data"

if(!dir.exists(here("sample_data"))){
  dir.create(here("sample_data"))
  
  # file.copy(stephanie_data_path, here("sample_data"), recursive=FALSE)
  file.copy(list.files(albert_data_path, full.names = TRUE), here("sample_data"), recursive=TRUE)
}

# verify the tumor files has been transferred. 
tumor_names <- list.files(here("sample_data"))[
                  !grepl("*.txt", list.files(here("sample_data")))]
print(tumor_names)
```

Next, we create the folder to store reference files (e.g. genome and cDNA) and quantification files (quantified counts) from salmon.  

```{r}
if(!dir.exists(here("salmon_files"))){
  dir.create(here("salmon_files"))
}

if(!dir.exists(here("salmon_quants"))){
  dir.create(here("salmon_quants"))
}
```

Create a list of paths for each cell which will be used when we quantify with salmon. 

```{r}
if(!file.exists(here("sample_data", "unique_cell_paths.txt"))){
  unique_cell_paths <- NULL
  for(tum in tumor_names){
    ids <- list.files(here("sample_data", tum))
    ids <- unique(stringr::str_sub(ids, end=-11))
    ids <- here("sample_data", tum, ids)
    unique_cell_paths <- c(unique_cell_paths, ids)
  }
  readr::write_lines(unique_cell_paths,
                     path = here("sample_data","unique_cell_paths.txt"))
}
```

# Quantification 

## Download files needed for salmon

We need to download 3 files:

1. `GRCh38.primary_assembly.genome.fa.gz` - nucleotide (DNA) sequences of the **GRCH38 primary genome assembly** (chromosomes and scaffolds -- i.e. unplaced scaffolds?)
2. `gencode.v32.transcripts.fa.gz` - nucleotide (DNA) sequences of **all transcripts** on reference chromosomes
3. `gencode.v32.annotation.gtf.gz` - gene annotation on the reference chromosomes (i.e. for humans, these are chromosomes 1 to 22, X, and Y), i.e. locations of genes and other information about the genes, gene structure
  * Gene transfer format (GTF) is a file format used to hold information about gene structure. It is a tab-delimited text format based on the general feature format (GFF), but contains some additional conventions specific to gene information.

Source: https://www.gencodegenes.org/human/

```{r}
# download GENCODE primary assembly fasta file
if(!file.exists(here("salmon_files", "GRCh38.primary_assembly.genome.fa.gz"))){
  tar_gz_file <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.primary_assembly.genome.fa.gz"
  download.file(tar_gz_file, 
                destfile = here("salmon_files", "GRCh38.primary_assembly.genome.fa.gz"), 
                method = "wget")
}

# download GENCODE transcripts fasta file
if(!file.exists(here("salmon_files", "gencode.v32.transcripts.fa.gz"))){
  tar_gz_file <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.transcripts.fa.gz"
  download.file(tar_gz_file, 
                destfile = here("salmon_files", "gencode.v32.transcripts.fa.gz"), 
                method = "wget")
}

# download GENCODE gtf file
if(!file.exists(here("salmon_files", "gencode.v32.annotation.gtf.gz"))){
  tar_gz_file <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.annotation.gtf.gz"
  download.file(tar_gz_file, 
                destfile = here("salmon_files", "gencode.v32.annotation.gtf.gz"), 
                method = "wget")
}
```


## Make pre-mRNA files for salmon

We need to create a set of pre-mRNA sequences to save to a FASTA file. 
This is needed because we have single nucleus RNA-seq data so we need to take into account the pre-mRNA and mRNA transcripts. 
This code kindly comes from [Charlotte Soneson](https://csoneson.github.io).

```{r}
suppressPackageStartupMessages({
  library(Biostrings)
  library(rtracklayer)
  library(GenomicFeatures)
  library(BSgenome)
})
```

### Combined GTF file

First, make a pre-mRNA GTF file (using the mRNA GTF file). Then combine it with the mRNA GTF file to get a GTF file for both pre-mRNA and mRNA.

```{r}
## Read gtf file and group exons by transcript
ingtf <- here("salmon_files","gencode.v32.annotation.gtf.gz")
txdb <- GenomicFeatures::makeTxDbFromGFF(ingtf)
grl <- GenomicFeatures::exonsBy(txdb, by = "tx", use.names = TRUE)

## Create the pre-mRNA coordinates by adding all introns to each transcript
grrange <- unlist(range(grl)) # 227462 rows
names(grrange) <- paste(names(grrange), "premrna", sep= ".")

## Save pre-mRNA gtf file
premrna_gtf <- here("salmon_files","gencode.v32.premrnaannotation.gtf")
rtracklayer::export(grrange, premrna_gtf)

## Combine two gtf files (pre-mRNA and mRNA) to be used in tximeta a bit later
unlist(grl) # 1372308 rows
tmp <- c(grrange, unlist(grl)) # 1599770 rows, pre-mRNA comes in front of mRNA

## example of a transcript with three exons (mRNA) or entire pre-mRNA region
tmp[grep("ENST00000456328.2", names(tmp)),]

## Save combined pre-mRNA and mRNA gtf file
preandmrna_gtf <- here("salmon_files","gencode.v32.preandmrnaannotation.gtf")
rtracklayer::export(tmp, preandmrna_gtf)
```

### Combined FASTA file

Now, for each entry in `grrange`, we extract the sequence from the corresponding `seqnames` entry in the `genome`.

**Note**: If you use Ensembl, there will be a discrepancy between the transcripts (e.g. ENST00000456328 vs ENST00000456328.1). 
Technically, it doesn't matter this point, but it will when we merge with the spliced (mRNA) transcripts.
So, if you do this, you will likely need to remove the version numbers before merging the pre-mRNA and mRNA fasta files. 
Or just use the Gencode annotation where the naming is consistent between the transcriptome fasta and the gtf.

```{r}
## Get genomic (DNA) sequence from FASTA file
genomefasta <- here("salmon_files", "GRCh38.primary_assembly.genome.fa.gz")

## Import FASTA file (DNA) with Biostrings::readDNAStringSet
genome <- Biostrings::readDNAStringSet(genomefasta)
names(genome) <- sapply(strsplit(names(genome), " "), .subset, 1) # creates chr1, etc

## Get the sequence of each pre-mRNA molecule
premrnaseq <- BSgenome::getSeq(x = genome, names = grrange)

## Save pre-mRNA sequences to fasta file 
premrna_fasta <- here("salmon_files", "GRCh38.premRNA.fa")
Biostrings::writeXStringSet(premrnaseq, filepath = premrna_fasta)
```

Let's check to make sure the pre-mRNA and the mRNA have the same number of transcripts.

```{bash}
cd /fastscratch/myscratch/shicks1/alsf-filbin/salmon_files
# cd /fastscratch/myscratch/akuo/alsf-filbin/salmon_files
zcat gencode.v32.transcripts.fa.gz | grep ">"  | wc -l # mRNA
grep ">" GRCh38.premRNA.fa | wc -l                     # pre-mRNA
```

Both have the same number of rows (or transcripts): 227462.
I think it's also interesting that one is almost 6GB and the other is 70M. 

Now we will put the pre-mRNA (`GRCh38.premRNA.fa`) and mRNA (`gencode.v32.transcripts.fa.gz`) fasta GENCODE files together. 

```{bash}
cd /fastscratch/myscratch/shicks1/alsf-filbin/salmon_files
# cd /fastscratch/myscratch/akuo/alsf-filbin/salmon_files
zcat gencode.v32.transcripts.fa.gz > gencode.v32.transcripts.fa
cat gencode.v32.transcripts.fa GRCh38.premRNA.fa > gencode.v32.preandmrna.fa
grep ">" gencode.v32.preandmrna.fa | wc -l
```

Ok there are 454924 (or = 227462*2) transcripts, as we expect. Now let's zip this file up. 
This takes a while (~30-40 mins). 

```{bash}
cd /fastscratch/myscratch/shicks1/alsf-filbin/salmon_files
# cd /fastscratch/myscratch/akuo/alsf-filbin/salmon_files
gzip gencode.v32.preandmrna.fa

# gzip the pre-mRNA only
gzip GRCh38.premRNA.fa
```

And clean up our space

```{bash}
cd /fastscratch/myscratch/shicks1/alsf-filbin/salmon_files
# cd /fastscratch/myscratch/akuo/alsf-filbin/salmon_files
rm gencode.v32.transcripts.fa
rm gencode.v32.transcripts.fa.gz
rm gencode.v32.annotation.gtf.gz
rm GRCh38.premRNA.fa
rm gencode.v32.premrnaannotation.gtf
rm GRCh38.primary_assembly.genome.fa.gz
```

We have made a GTF file (`gencode.v32.preandmrnaaannotation.gtf`) and a FASTA file (`gencode.v32.preandmrna.fa`) for the combined pre-mRNA and mRNA.

### Alternative way: extracting introns

```{r}
# Modified from https://github.com/csoneson/rna_velocity_quant/blob/master/scripts/extractIntronSeqs.R
#' Extract intron sequences
#'
#' @param gtf The path to a gtf file
#' @param genome A \code{DNAStringSet} object with the genome sequence
#' @param type Either 'collapse' or 'separate'
#' @param flanklength The length of the exonic flanking sequence
#' @param joinOverlappingIntrons Whether overlapping intron sequences (after adding 
#'   the flanking sequence) should be joined into a single intron
#'
#' @return A \code{DNAStringSet} object with intronic sequences
#' 
extractIntronSeqs <- function(gtf, genome, type = "collapse", flanklength = 90,
                              joinOverlappingIntrons = FALSE) {
  ## Construct TxDb from gtf file
  txdb <- GenomicFeatures::makeTxDbFromGFF(gtf, format = "gtf")

  if (type == "separate") {
    ## Group exons by transcript
    grl <- GenomicFeatures::exonsBy(txdb, by = "tx", use.names = TRUE)
  } else if (type == "collapse") {
    ## Group exons by gene
    grl <- GenomicFeatures::exonsBy(txdb, by = "gene")

    ## Collapse the exons of each gene
    grl <- GenomicRanges::reduce(grl)
  } else {
    stop("Unknown 'type' argument")
  }

  ## Get introns as the set difference between the range and the exons,
  ## for each transcript/gene
  ## Here, the order of the introns doesn't really matter, since they
  ## will be considered separately (not joined together into a transcript)
  grl <- BiocGenerics::setdiff(range(grl), grl)

  ## Add flanking region
  grl <- grl + flanklength

  if (joinOverlappingIntrons) {
    ## If two (introns + flanklength) overlap, join them
    grl <- GenomicRanges::reduce(grl)
  }
  
  gr <- unlist(grl)

  ## Add -I{X} to names
  names(gr) <- gsub("\\-I\\.", "-I", make.unique(paste0(names(gr), "-I")))

  ## Get sequence
  gs <- BSgenome::getSeq(x = genome, names = gr)
  
  ## Manually set names of extracted sequences
  stopifnot(all(width(gs) == width(gr)))
  names(gs) <- names(gr)
  
  return(list(gs = gs,
              gr = gr))
}
```

```{r}
ingtf <- here("salmon_files","gencode.v32.annotation.gtf.gz")
genomefasta <- here("salmon_files", "GRCh38.primary_assembly.genome.fa.gz")
if(!exists("genome")){
  genome <- Biostrings::readDNAStringSet(genomefasta)
}

# Get intron sequence and gtf
intronseq <- extractIntronSeqs(gtf = ingtf, genome = genome, 
                  type = "collapse", flanklength = 90,
                  joinOverlappingIntrons = FALSE)

# Save intron sequences to fasta file
intron_fasta <- here("salmon_files", "GRCh38.intron.fa")
Biostrings::writeXStringSet(intronseq$gs, filepath = intron_fasta)

# Save intron gtf file
intron_gtf <- here("salmon_files","gencode.v32.intronannotation.gtf")
rtracklayer::export(intronseq$gr, intron_gtf)
```

## Install and build salmon index 

This part will have to be done for each user. 
I installed the salmon 1.0.0 binary in my home directory here `/users/shicks1/src/`. 

To install salmon v1.0.0: 
```{bash}
cd /users/shicks1/src/

wget https://github.com/COMBINE-lab/salmon/releases/download/v1.0.0/salmon-1.0.0_linux_x86_64.tar.gz
tar xzvf salmon-1.0.0_linux_x86_64.tar.gz
rm salmon-1.0.0_linux_x86_64.tar.gz
```

Also, make sure this is in the `.bash_profile` file
```{bash}
PATH=$PATH:/users/shicks1/src/salmon-latest_linux_x86_64/bin
```

You can check to make sure salmon has been upgraded correctly using `salmon -h` inside terminal (or help with specific parts of using salmon using e.g. `salmon index -h` for help with the index step). 

OK, we are ready to use salmon. 
The `-t` argument is the input transcripts file. 
The `-i` argument is the index file to create. 
The `--keepDuplicates` argument forces all duplicate transcripts (for example, multiple unspliced transcript of the same gene that are identical for example) that appear in the input will be retained and quantified separately. 
If you keep the duplicates they will be assigned identical expression levels since salmon can’t tell them apart. 
When you aggregate on the gene level, this will not make a difference any more. 
Therefore, I do not keep the duplicates as we are interested in gene level aggregation. 
The `--gencode` flag will handle the composite fasta headers in GENCODE transcript fasta files and split the transcript name at the first '|' character. 
The `--threads` argument says how many threads to use when building the index. 

There is a script `build-index-salmon.sh` in the `/scripts` folder that was used to run this code with 4 cores. The index is built from the combined FASTA file.

## Run salmon 

We will now use the index created by `build-index-salmon.sh`.
See the `scripts/run-salmon.sh` file or the `scripts/run-salmon-parallel.sh` file in the terminal. They do the same thing, but the latter will run salmon in parallel and be much faster.

## Set up the tximeta package 

Next, we use the `tximeta` package to create a SummarizedExperiment object.
If we try to run `tximeta(coldata)`, that will fail because `tximeta` won't be able to detect automatically what the reference transcriptome is. 
So, here we create our own [linked transcriptome](https://bioconductor.org/packages/release/bioc/vignettes/tximeta/inst/doc/tximeta.html#linked_transcriptomes) from the combined GTF file.
This step is not necessary if your index was only the mRNA transcripts.

A `linkedTxome` records key information about the sources of the transcript FASTA files, and the location of the relevant GTF file. 
It also records the checksum of the transcriptome that was computed by Salmon during the index step.

```{r}
suppressPackageStartupMessages({
  library(tximeta)
})

# create linkedTranscriptome for combined list of pre-mRNA and mRNA
index_dir <- here("salmon_files", "gencode.v32_salmon-index-v1.0.0")
fasta_path <-  here("salmon_files", "gencode.v32.preandmrna.fa.gz")
gtf_path <-  here("salmon_files", "gencode.v32.preandmrnaaannotation.gtf") 

# create linkedTranscriptome for pre-mRNA only
index_dir <- here("salmon_files", "gencode.v32_salmon-index-v1.0.0-premRNA")
fasta_path <-  here("salmon_files", "GRCh38.premRNA.fa.gz")
gtf_path <-  here("salmon_files", "gencode.v32.premrnaannotation.gtf") 

json_file <- here("salmon_files", paste0(basename(index_dir), ".json"))
makeLinkedTxome(indexDir=index_dir,
                source="other", organism="Homo sapiens",
                release="other", genome="GRCh38",
                fasta=fasta_path, gtf=gtf_path,
                jsonFile=json_file) # this command will add the index to the cache automatically
```


