---
title: qc and analysis of single nuc-seq data (alsf-filbin)
author: Stephanie Hicks and Albert Kuo
output: 
  html_document:
    code_folding: "hide"
    toc: TRUE
    toc_float: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

```{bash eval = F}
# Commands to start an interactive session on the JHPCE cluster
qrsh -l mem_free=20G,h_vmem=20G
module load conda_R
cd /fastscratch/myscratch/akuo/alsf-filbin
R
```

```{r eval = F}
library(here)
dataset = "preandmrna" # options = "transcripts", "preandmrna" (might need more memory for preandmrna)
```

# FASTQ read length distribution

The average read length from 1 FASTQ file is 25 bp.

```{bash eval = F}
cat BT1179Nuc-P4-H12_R2.fastq.gz | awk '{if(NR%4==2) print length($1)}' | sort -n | uniq -c > read_length_BT1179Nuc.txt
```

```{r eval = F}
reads <- read.csv(here("sample_data", "read_length_BT1179Nuc.txt"), sep="", header=FALSE)
plot(reads$V2,reads$V1,type="l",xlab="read length",ylab="occurences",col="blue")
```

# Read in quant files

## SE object

We read in the quantification files using `tximeta` and save the SummarizedExperiment object.

* https://github.com/csoneson/rna_velocity_quant/blob/488b9f6c479201c99b588586e27b31a26faacaba/scripts/sce_helpers.R 

```{r eval = F}
# list tumor names
tumor_names <- list.files(here("sample_data"))[
                  !grepl("*.txt", list.files(here("sample_data")))]

# Create paths for tximeta 

unique_sf_paths <- NULL
for(tum in tumor_names){
  ids <- list.files(here("sample_data", tum))
  ids <- unique(stringr::str_sub(ids, end=-13))
  if(!identical(ids, character(0))){
    ids <- here(paste0("salmon_quants_", dataset), paste0(ids, "_quant"), "quant.sf")
  }
  unique_sf_paths <- c(unique_sf_paths, ids)
}

unique_sf_ids <- NULL
for(tum in tumor_names){
  ids <- list.files(here("sample_data", tum))
  ids <- unique(stringr::str_sub(ids, end=-13))
  unique_sf_ids <- c(unique_sf_ids, ids)
}

coldata <- data.frame(files = unique_sf_paths, names = unique_sf_ids, stringsAsFactors = FALSE)
```

We also need to use the linkedTxome object to use `tximeta` properly, i.e. `rowRanges(se)` won't be `NULL` and tximeta will be able to match the transcripts to the genes for us. **Note: still doesn't work**

```{r eval = F}
suppressPackageStartupMessages({
  library(tximeta)
  library(BiocFileCache)
})

# check if linkedTxome is already in the cache
bfcloc <- getTximetaBFC()
bfc <- BiocFileCache(bfcloc)
bfcinfo(bfc)

# if not, load linkedTxome json file
json_file <- here("salmon_files", "gencode.v32_salmon-index-v1.0.0.json")
loadLinkedTxome(json_file)
```

The only way I've figured out how to make this work for now is with `skipMeta = TRUE`.

```{r eval = F}
se_file_name = here(paste0("salmon_quants_", dataset), paste0("se_", dataset, ".rds"))

# coldata = coldata[1:2, ] # for testing
if(!file.exists(se_file_name)){
  # Create SummarizedExperiment object
  if(dataset == "preandmrna")
    se <- tximeta(coldata, skipMeta = TRUE) # Takes a couple of minutes, file size = 800 MB
  else if (dataset == "transcripts")
    se <- tximeta(coldata, type = "salmon")
  # se <- tximeta(coldata, ignoreAfterBar = TRUE)
  saveRDS(se, se_file_name)
} else {
  se = readRDS(se_file_name) # Takes a couple of seconds
}

# Check SummarizedExperiment object
colData(se)
assayNames(se)
rowRanges(se) # empty
```


## SCE object

We add tx2gene to rowData and convert the SummarizedExperiment object to a SingleCellExperiment object.

A SingleCellExperiment class is derived from the SummarizedExperiment class. The most important change is the addition of a new slot called `reducedDims`. Read more [here](https://osca.bioconductor.org/data-infrastructure.html#a-brief-recap-from-se-to-sce).

```{r eval = F}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(SingleCellExperiment)
  library(DESeq2)
  library(scater)
})

# Read in SE file
se_file_name = here(paste0("salmon_quants_", dataset), paste0("se_", dataset, ".rds"))
se = readRDS(se_file_name) # Takes a couple of seconds

# Add rowData manually (gene names)
library(readr)
if(dataset == "transcripts"){
  tx2gene = read_tsv(here("salmon_files", "gencode.v32.annotation.tx2gene.txt"))
} else if(dataset == "preandmrna"){
  tx2gene = read_tsv(here("salmon_files", "gencode.v32.preandmrna.tx2gene.txt"))
}

match_rows = match(rownames(se), tx2gene$transcript_id)
tx2gene = tx2gene[match_rows, ]
rowData(se) = tx2gene

# Coercion from SummarizedExperiment
sce = as(se, "SingleCellExperiment")
# you can access counts by assay(sce, "counts") or counts(sce)
# you can add a new entry to assays slot by assay(sce, "counts_new") = dat_new

# Save SCE file
sce_file_name = here(paste0("salmon_quants_", dataset), paste0("sce_", dataset, ".rds"))
saveRDS(sce, sce_file_name)
```

# [Unrevised] Quality Control

There are a couple of [QC metrics](https://osca.bioconductor.org/quality-control.html#choice-of-qc-metrics) to identify low-quality cells:

1. Using counts, i.e. cells with (a) a small library size (total sum of counts) `low_lib_size` or (b) few expressed endogeneous genes (nonzero counts for those genes) `low_n_features`
2. Using "spike-in transcripts", i.e. any enrichment of spike-in transcripts (higher proportion) 
3. Using the mitochondrial genome, i.e. any enrichment of reads in the mitochondrial genome is indicative of loss of cytoplasmic RNA

I will only do (1) for now.

```{r eval = F}
library(scater)

# Read in SCE object
sce_file_name = here(paste0("salmon_quants_", dataset), paste0("sce_", dataset, ".rds"))
sce = readRDS(sce_file_name) # Takes a couple of seconds

# Compute quality control metrics:
# sum is the total count for each cell
# detected contains the number of detected genes (actually transcripts for our data)
df = perCellQCMetrics(sce)
df

# Find outliers with low library sizes (LibSize) and few detected features (n_features)
reasons = quickPerCellQC(df) # DataFrame of logical values
colSums(as.matrix(reasons))

# Discard outliers
filtered = sce[, !reasons$discard]
dat_filtered = counts(filtered) # 226608 x 1018 for mrna, 221988 x 1090 for premrna
sce_filtered = SingleCellExperiment(assays = list(counts = dat_filtered))

filtered_file_name = here(paste0("salmon_quants_", dataset), paste0("sce_filtered_", dataset, ".rds"))
saveRDS(sce_filtered, filtered_file_name)
```

Diagnostic plots: https://osca.bioconductor.org/quality-control.html#quality-control-plots 

# [Unrevised] Transform read counts

```{r eval = F}
library(here)
library(ggplot2)
library(dplyr)
library(tidyr)
library(SingleCellExperiment)
dataset = "transcripts" # options = "preandmrna", "transcripts"
```

## Convert read counts to pseudo-UMIs

Run `compute-mle.sh` to get mle sig ("shape") parameter first. There will be a parameter for every cell. It will take about 30 minutes to an hour.

```{r eval = F}
# Load filtered counts
filtered_file_name = here(paste0("salmon_quants_", dataset), paste0("sce_filtered_", dataset, ".rds"))
if(file.exists(filtered_file_name)){
  sce_filtered = readRDS(filtered_file_name)
  dat_filtered = counts(sce_filtered)
}

# Read in mle parameters
mle_file_names = list.files(here(paste0("mle_results_", dataset)), full.names = TRUE)
mle_results = lapply(mle_file_names, readRDS)
sig_vec = sapply(mle_results, function(r) r["sig"])
mu_vec = sapply(mle_results, function(r) r["mu"])

# Plot sig ("shape") distribution
ggplot(tibble(sig = sig_vec), aes(x = sig)) +
  geom_histogram(bins = 30) +
  labs(title = "Distribution of sig parameter",
       x = "sig",
       y = "Number of cells")
```

```{r eval = F}
source(here("scripts", "quminorm.R"))

# Convert to pseudo-UMIs
umi_file_name = here(paste0("salmon_quants_", dataset), paste0("sce_umi_", dataset, ".rds"))
if(file.exists(umi_file_name)){
  dat_umi = readRDS(umi_file_name)
} else {
  dat_umi = quminorm_matrix(dat_filtered, shape = median(sig_vec)) # Takes several minutes (20~30 min). In the paper, they use the mode, but I will use the median since it is not clear what mode means for continuous values. The median for mRNA is 3.16. The median for pre-mRNA is 3.28. 
  dat_umi = dat_umi[!(rowSums(dat_umi) == 0), ] # Remove transcripts with all zeros
  sce_umi = SingleCellExperiment(assays = list(counts = dat_umi))
  saveRDS(sce_umi, umi_file_name)
}
```


# Exploratory Plots

```{r}
library(here)
library(tidyverse)
library(SingleCellExperiment)
library(scattermore)
library(tictoc)
```

## Number of counts

Plot number of counts mapping to unspliced ("pre-mRNA") vs spliced ("mRNA") under different pipelines.

```{r eval = F}
sce_preandmrna = readRDS(here("salmon_quants_preandmrna", "sce_preandmrna.rds"))
sce_transcripts = readRDS(here("salmon_quants_transcripts", "sce_transcripts.rds"))
```

```{r eval = F}
# Counts matrix for preandmrna pipeline
counts_preandmrna = counts(sce_preandmrna)
dim(counts_preandmrna) # 423709 x 1152
# Counts matrix for transcripts pipeline
counts_transcripts = counts(sce_transcripts)
dim(counts_transcripts) # 226608 x 1152

# Number of reads mapped to unspliced ("pre-mRNA") for preandmrna pipeline
sum_preandmrna_unspliced = colSums(counts_preandmrna[grep("unspliced", rownames(counts_preandmrna)), ])
# Number of reads mapped to spliced ("mRNA") for preandmrna pipeline
sum_preandmrna_spliced = colSums(counts_preandmrna[-grep("unspliced", rownames(counts_preandmrna)), ])
# Number of reads mapped to spliced ("mRNA") for transcripts pipeline
sum_transcripts_spliced = colSums(counts_transcripts)
```

```{r eval = F}
# Create summary table
sum_counts_tb = tibble(sample = colnames(counts_preandmrna),
                       preandmrna_unspliced = sum_preandmrna_unspliced,
                       preandmrna_spliced = sum_preandmrna_spliced,
                       transcripts_spliced = sum_transcripts_spliced) %>%
  mutate(preandmrna_both = preandmrna_unspliced + preandmrna_spliced)

sum_counts_tb = sum_counts_tb %>% 
  pivot_longer(-sample, names_to = c("source")) %>%
  mutate(pipeline = ifelse(grepl("preandmrna", source), "preandmrna", "transcripts"))
saveRDS(sum_counts_tb, file = here("data", "sum_counts_tb.rds"))
```

```{r, fig.width = 7, fig.height = 7}
# All reads and samples
sum_counts_tb = readRDS(here("data", "sum_counts_tb.rds"))
sum_counts_1_tb = sum_counts_tb %>%
  filter(source != "preandmrna_both")
sum_counts_1_tb_medians = sum_counts_1_tb %>%
  group_by(source) %>%
  summarize(medians = median(value))

sum_counts_1_tb %>%
  ggplot(aes(x = source, y = value)) +
  geom_boxplot(aes(color = pipeline)) +
  geom_text(data = sum_counts_1_tb_medians, aes(x = source, y = medians, label = round(medians)), 
              position = position_dodge(width = 0.8), size = 3, vjust = -0.5) +
  labs(title = "Number of reads mapped to spliced vs unspliced",
       y = "Number of reads") +
  theme_bw()

sum_counts_2_tb = sum_counts_tb %>%
  filter(source %in% c("preandmrna_both", "transcripts_spliced"))
sum_counts_2_tb_medians = sum_counts_2_tb %>%
  group_by(source) %>%
  summarize(medians = median(value))
sum_counts_2_tb %>%
  ggplot(aes(x = source, y = value)) +
  geom_boxplot(aes(color = pipeline)) +
  geom_text(data = sum_counts_2_tb_medians, aes(x = source, y = medians, label = round(medians)), 
              position = position_dodge(width = 0.8), size = 3, vjust = -0.5) +
  labs(title = "Number of reads mapped overall under different pipelines",
       y = "Number of reads") +
  theme_bw()
```

```{r, fig.width = 7, fig.height = 7}
# preandmrna_spliced vs transcripts_spliced
sum_counts_tb %>%
  mutate(patient = gsub("-.*", "", sample)) %>%
  select(-pipeline) %>%
  pivot_wider(names_from = "source", values_from = "value") %>%
  ggplot(aes(x = log(transcripts_spliced), y = log(preandmrna_spliced), color = patient)) +
  geom_point(size = 0.5) +
  geom_abline(intercept = 0, slope = 1) +
  labs(title = "Number of reads mapped to spliced transcripts under different pipelines",
       subtitle = "Each point is a different cell") +
  theme_bw()

# preandmrna_both vs transcripts_spliced
sum_counts_tb %>%
  mutate(patient = gsub("-.*", "", sample)) %>%
  select(-pipeline) %>%
  pivot_wider(names_from = "source", values_from = "value") %>%
  ggplot(aes(x = log(transcripts_spliced), y = log(preandmrna_both), color = patient)) +
  geom_point(size = 0.5) +
  geom_abline(intercept = 0, slope = 1) +
  labs(title = "Number of reads mapped overall under different pipelines",
       subtitle = "Each point is a different cell") +
  theme_bw()
```

```{r}
# Grouped by patient
sum_counts_tb %>%
  mutate(patient = gsub("-.*", "", sample)) %>%
  ggplot(aes(x = source, y = value, color = pipeline)) +
  geom_boxplot() +
  facet_wrap(~ patient, nrow = 1) +
  labs(title = "All reads and samples, grouped by patient") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

## MA plot

### Aggregate counts at gene level

```{r eval = F}
library(here)
library(tidyverse)
library(SingleCellExperiment)
library(scattermore)
library(tictoc)
```

```{r eval = F}
sce_preandmrna = readRDS(here("salmon_quants_preandmrna", "sce_preandmrna.rds"))
sce_transcripts = readRDS(here("salmon_quants_transcripts", "sce_transcripts.rds"))
```

```{r eval = F}
# Counts matrix for preandmrna pipeline
counts_preandmrna = counts(sce_preandmrna)
dim(counts_preandmrna) # 423709 x 1152

# Counts matrix for transcripts pipeline
counts_transcripts = counts(sce_transcripts)
dim(counts_transcripts) # 226608 x 1152
```

```{r eval = F}
# Sum over gene for preandmrna pipeline
counts_bygene_preandmrna = cbind(counts_preandmrna, 
                                  rowData(sce_preandmrna) %>%
                                    as_tibble() %>%
                                    select(gene_id, gene_type)) %>%
  group_by(gene_id, gene_type) %>%
  summarize_all(sum) %>%
  ungroup()

saveRDS(counts_bygene_preandmrna, here("salmon_quants_preandmrna", "counts_bygene_preandmrna.rds"))

# Sum over gene for transcripts pipeline
counts_bygene_transcripts = cbind(counts_transcripts, 
                                  rowData(sce_transcripts) %>%
                                    as_tibble() %>%
                                    select(gene_id, gene_type)) %>%
  group_by(gene_id, gene_type) %>%
  summarize_all(sum) %>%
  ungroup()

saveRDS(counts_bygene_transcripts, here("salmon_quants_transcripts", "counts_bygene_transcripts.rds"))
```

### Plot

```{r}
counts_bygene_preandmrna = readRDS(here("salmon_quants_preandmrna", "counts_bygene_preandmrna.rds"))
dim(counts_bygene_preandmrna)
counts_bygene_transcripts = readRDS(here("salmon_quants_transcripts", "counts_bygene_transcripts.rds"))
dim(counts_bygene_transcripts)
```

Plot for one cell only

```{r}
# Bind columns
counts_bygene_onecell_preandmrna = counts_bygene_preandmrna %>%
  select(c("gene_id", "gene_type", "BT1179Nuc-P1-A01")) %>%
  dplyr::rename("BT1179Nuc-P1-A01-preandmrna" = "BT1179Nuc-P1-A01")

counts_bygene_onecell_transcripts = counts_bygene_transcripts %>%
  select(c("gene_id", "gene_type", "BT1179Nuc-P1-A01")) %>%
  dplyr::rename("BT1179Nuc-P1-A01-transcripts" = "BT1179Nuc-P1-A01")

counts_bygene_onecell = counts_bygene_onecell_preandmrna %>%
  inner_join(., counts_bygene_onecell_transcripts, by = c("gene_id", "gene_type"))

# Table of most frequent gene types
counts_bygene_onecell %>% 
  group_by(gene_type) %>% summarize(num_reads_mapped = sum(`BT1179Nuc-P1-A01-preandmrna`)) %>% 
  arrange(-num_reads_mapped)

# Plot
counts_bygene_onecell %>%
  mutate(m = log(`BT1179Nuc-P1-A01-preandmrna`) - log(`BT1179Nuc-P1-A01-transcripts`),
         a = (log(`BT1179Nuc-P1-A01-preandmrna`) + log(`BT1179Nuc-P1-A01-transcripts`)/2)) %>%
  mutate(gene_type_color = ifelse(gene_type %in% c("protein_coding", "lncRNA",
                                                   "transcribed_unprocessed_pseudogene"),
                                  gene_type, "all_other_types")) %>%
  ggplot(aes(x = a,
             y = m)) + 
  #stat_density2d(aes(fill = ..density..^0.3), geom = "tile", contour = FALSE, n = 200) +
  #scale_fill_continuous(low = "white", high = "#1C5E7E") + 
  geom_hline(yintercept = 0, color = "red") +
  geom_smooth() +
  geom_point(alpha = 0.3) +
  facet_wrap(~ gene_type_color) +
  labs(x = "A",
       y = "M") +
  theme_bw()
```

Plot for average across all cells

```{r}
# Bind columns
counts_bygene_avg_preandmrna = counts_bygene_preandmrna %>%
  mutate_at(vars(-c("gene_id", "gene_type")), ~(./sum(.))) %>% # normalize by library size
  mutate(avg_preandmrna = rowMeans(counts_bygene_preandmrna %>% select(-c("gene_id", "gene_type")))) %>%
  select(c("gene_id", "gene_type", "avg_preandmrna"))

counts_bygene_avg_transcripts = counts_bygene_transcripts %>%
  mutate_at(vars(-c("gene_id", "gene_type")), ~(./sum(.))) %>% # normalize by library size
  mutate(avg_transcripts = rowMeans(counts_bygene_transcripts %>% select(-c("gene_id", "gene_type")))) %>%
  select(c("gene_id", "gene_type", "avg_transcripts"))

counts_bygene_avg = counts_bygene_avg_preandmrna %>%
  inner_join(., counts_bygene_avg_transcripts, by = c("gene_id", "gene_type"))

# Table of most frequent gene types
counts_bygene_avg %>% 
  group_by(gene_type) %>% summarize(sum = sum(avg_preandmrna)) %>% 
  arrange(-sum)

# Plot
counts_bygene_avg %>%
  mutate(m = log(`avg_preandmrna`) - log(`avg_transcripts`),
         a = (log(`avg_preandmrna`) + log(`avg_transcripts`)/2)) %>%
  mutate(gene_type_color = ifelse(gene_type %in% c("protein_coding", "lncRNA",
                                                   "transcribed_unprocessed_pseudogene"),
                                  gene_type, "all_other_types")) %>%
  ggplot(aes(x = a,
             y = m)) + 
  #stat_density2d(aes(fill = ..density..^0.3), geom = "tile", contour = FALSE, n = 200) +
  #scale_fill_continuous(low = "white", high = "#1C5E7E") + 
  geom_hline(yintercept = 0, color = "red") +
  geom_smooth() +
  geom_point(alpha = 0.01) +
  facet_wrap(~ gene_type_color) +
  labs(x = "A",
       y = "M") +
  theme_bw()

# smoothscatter version
# smoothScatter(x = dat_rowmeans$a, y = dat_rowmeans$m,
#               xlab = "A",
#               ylab = "M")
```

## [Unrevised] Comparison to parametric distributions

### pre-mRNA

First, I make all the column sums the same by scaling. The count in row $i$ and column $j$ is transformed as $x_{ij} = x_{ij}*\sum_i x_{ij}/median_j(\sum_i x_{ij})$. 

```{r pre-mRNA_distribution}
summary(colSums(dat_prem))

# Apply transformation
n = round(median(colSums(dat_prem)))
dat_prem = apply(dat_prem, MARGIN = 2, function(x) x*n/sum(x))

summary(colSums(dat_prem))
```

I plot the mean and variance for every row (transcript). Under a poisson distribution, they should be the same.

```{r}
plot_meanvar = function(dat_sub){
  # estimate lambdas and variances for every transcript
  means = rowMeans(dat_sub)
  vars = apply(dat_sub, MARGIN = 1, var)
  
  # variance = mean under Poisson distribution
  tibble(means, vars) %>%
    ggplot(aes(x = log(means), y = log(vars))) +
    geom_point(alpha = 0.4) +
    geom_abline(intercept = 0, slope = 1)
}
```

```{r}
plot_meanvar(dat_prem) +
  labs(title = "pre-mRNA")
```

I first compute the average expression for every row (x-axis) $\hat{\mu}_i$ and the empirical $P(X_i=0)$, which is the probability that for a given transcript $i$, the count is 0. 

I then compute what $P(X_i=0)$ would be under the model assumptions of binomial or poisson, using parameters estimated from the data. In particular, for a $Binom(n, p_i)$, $n$ is the median number of total counts of cells and $p_i$ is the mean proportion of counts that were in gene $i$.

```{r}
# Function to plot P(X_i = 0) against average expression level mu_i
plot_prob = function(dat_sub){
  n = round(median(colSums(dat_sub)))
  means = rowMeans(dat_sub)
  vars = apply(dat_sub, MARGIN = 1, var)
  
  # empirical P(X_i = 0)
  emp_probs_0 = apply(dat_sub, MARGIN = 1, function(r) sum(r==0)/ncol(dat_sub))
  plot_dt = tibble(means, emp_probs_0)
  
  emp_props = rowMeans(dat_sub/colSums(dat_sub))
  # Model P(X_i = 0) under Binomial
  binom_probs_0 = dbinom(x = 0, size = n, prob = emp_props)
  # Model P(X_i = 0) under Poisson
  poiss_probs_0 = dpois(x = 0, lambda = n*emp_props)
  # Model P(X_i = 0) under Negative Binomial
  # Estimate size/dispersion parameter
  model = lm(vars ~ 1*means + I(means^2) + 0, tibble(means, vars))
  phi = 1/coef(model)["I(means^2)"]
  nbinom_probs_0 = dnbinom(x = 0, size = phi, mu = n*emp_props) 
  
  # Tibble for plot
  plot_lines_dt = tibble(means = means,
                         binomial = binom_probs_0,
                         poisson = poiss_probs_0,
                         nbinomial = nbinom_probs_0) %>%
    pivot_longer(-means, names_to = "model", values_to = "probs_0")
  
  # Plot
  plt = plot_lines_dt %>%
    ggplot(aes(x = log(means), y = probs_0)) +
    geom_point(data = plot_dt, aes(x = log(means), y = emp_probs_0), alpha = 0.4) + # Add data points
    geom_line(aes(color = model),
              size = 1.5) + # Add lines for models
    labs(x = "Average expression level log(E(X_i))",
         y = "P(X_i = 0)") +
    theme(text = element_text(size = 15))
  
  # Return object
  out = list(plot = plt,
             lines_dt = plot_lines_dt)
  return(out)
}
```

```{r}
# Plot P(X_i = 0) against average expression level
prob_out_prem = plot_prob(dat_prem)
prob_out_prem$plot +
  labs(title = "pre-mRNA")
```


### mRNA

```{r mRNA_distribution}
summary(colSums(dat_mrna))

# Apply transformation
n = round(median(colSums(dat_mrna)))
dat_mrna = apply(dat_mrna, MARGIN = 2, function(x) x*n/sum(x))

summary(colSums(dat_mrna))
```

```{r}
# Plot mean against variance
plot_meanvar(dat_mrna) +
  labs(title = "mRNA")
```

```{r}
# Plot P(X_i = 0) against average expression level
prob_out_mrna = plot_prob(dat_mrna)
prob_out_mrna$plot +
  labs(title = "mRNA")
```

## PCA & Clustering

```{r}
library(scran)
library(scater)
library(BiocSingular)
# library(factoextra)
library(tictoc)
library(glmpca)
# source(here("scripts", "functions_genefilter.R"))
```

### preandmrna pipeline

#### BT1179Nuc 

Doing PCA for the cells from one patient only.

```{r}
# Turn counts by gene into sce
counts = counts_bygene_preandmrna %>% select(starts_with("BT1179Nuc")) %>% as.matrix()
sce_bygene_preandmrna = SingleCellExperiment(assays = list(counts = counts),
                                             rowData = counts_bygene_preandmrna %>% select(-c("gene_id", "gene_type")))

# Normalize and take log
clust = quickCluster(sce_bygene_preandmrna)
sce_bygene_preandmrna = computeSumFactors(sce_bygene_preandmrna, clusters = clust)
sce_bygene_preandmrna = logNormCounts(sce_bygene_preandmrna)

set.seed(1)
tic("approx PCA")
sce_bygene_preandmrna = runPCA(sce_bygene_preandmrna, exprs_values = "logcounts", ntop = ncol(sce_bygene_preandmrna), BSPARAM = IrlbaParam())
pc_bygene_preandmrna = list(x = reducedDim(sce_bygene_preandmrna, "PCA"))
toc()

# Plot PC
pc = pc_bygene_preandmrna
tumor_labels = gsub("-.*$", "", colnames(sce_bygene_preandmrna))
lib_size = colSums(counts(sce_bygene_preandmrna))
prop_mapped = 1 - colMeans(!counts(sce_bygene_preandmrna))

# By library counts
data.frame(PC1 = pc$x[,1], PC2 = pc$x[,2], lib_size = lib_size) %>%
  ggplot(aes(PC1, PC2, color = lib_size)) +
  scale_colour_gradientn(colours = terrain.colors(10), trans = "sqrt") +
  geom_point(alpha = 0.5) +
  theme_bw()

data.frame(PC2 = pc$x[,2], PC3 = pc$x[,3], lib_size = lib_size) %>%
  ggplot(aes(PC2, PC3, color = lib_size)) +
  scale_colour_gradientn(colours = terrain.colors(10), trans = "sqrt") +
  geom_point(alpha = 0.5) +
  theme_bw()

# By proportion of zeroes (genes with no reads mapped)
data.frame(PC1 = pc$x[,1], PC2 = pc$x[,2], prop_mapped = prop_mapped) %>%
  ggplot(aes(PC1, PC2, color = prop_mapped)) +
  scale_colour_gradientn(colours = terrain.colors(10)) +
  geom_point(alpha = 0.5) +
  theme_bw()

data.frame(PC2 = pc$x[,2], PC3 = pc$x[,3], prop_mapped = prop_mapped) %>%
  ggplot(aes(PC2, PC3, color = prop_mapped)) +
  scale_colour_gradientn(colours = terrain.colors(10)) +
  geom_point(alpha = 0.5) +
  theme_bw()

# By patient
data.frame(PC1 = pc$x[,1], PC2 = pc$x[,2], patient = tumor_labels) %>%
  ggplot(aes(PC1, PC2, color = patient)) +
  geom_point(alpha = 0.5) +
  theme_bw()

data.frame(PC1 = pc$x[,1], PC3 = pc$x[,3], patient = tumor_labels) %>%
  ggplot(aes(PC1, PC3, color = patient)) +
  geom_point(alpha = 0.5) +
  theme_bw()
```


```{r, fig.height = 5, fig.width = 5}
# K-means clustering
set.seed(1)

# Elbow plot with SSE
sse = c()
for(k in 1:10){
  clust.kmeans <- kmeans(reducedDim(sce_bygene_preandmrna, "PCA"), centers = k)
  sse = c(sse, clust.kmeans$tot.withinss)
}
plot(sse)

clust.kmeans <- kmeans(reducedDim(sce_bygene_preandmrna, "PCA"), centers = 6)
colLabels(sce_bygene_preandmrna) <- factor(clust.kmeans$cluster)
plotReducedDim(sce_bygene_preandmrna, "PCA", colour_by="label")
```


### transcripts pipeline

#### BT1179Nuc

```{r}
# Turn counts by gene into sce
counts = counts_bygene_transcripts %>% select(starts_with("BT1179Nuc")) %>% as.matrix()
counts = counts[, colSums(counts) != 0] # temporary quality control
sce_bygene_transcripts = SingleCellExperiment(assays = list(counts = counts),
                                              rowData = counts_bygene_transcripts %>% select(-c("gene_id", "gene_type")))

# Normalize and take log
clust = quickCluster(sce_bygene_transcripts)
sce_bygene_transcripts = computeSumFactors(sce_bygene_transcripts, clusters = clust)
sce_bygene_transcripts = logNormCounts(sce_bygene_transcripts)

set.seed(1)
tic("approx PCA")
sce_bygene_transcripts = runPCA(sce_bygene_transcripts, exprs_values = "logcounts", ntop = ncol(sce_bygene_transcripts), BSPARAM = IrlbaParam())
pc_bygene_transcripts = list(x = reducedDim(sce_bygene_transcripts, "PCA"))
toc()

# Plot PC
pc = pc_bygene_transcripts
tumor_labels = gsub("-.*$", "", colnames(sce_bygene_transcripts))
lib_size = colSums(counts(sce_bygene_transcripts))
prop_mapped = 1 - colMeans(!counts(sce_bygene_transcripts))

# By library counts
data.frame(PC1 = pc$x[,1], PC2 = pc$x[,2], lib_size = lib_size) %>%
  ggplot(aes(PC1, PC2, color = lib_size)) +
  scale_colour_gradientn(colours = terrain.colors(10), trans = "sqrt") +
  geom_point(alpha = 0.5) +
  theme_bw()

data.frame(PC2 = pc$x[,2], PC3 = pc$x[,3], lib_size = lib_size) %>%
  ggplot(aes(PC2, PC3, color = lib_size)) +
  scale_colour_gradientn(colours = terrain.colors(10), trans = "sqrt") +
  geom_point(alpha = 0.5) +
  theme_bw()

# By proportion of zeroes (genes with no reads mapped)
data.frame(PC1 = pc$x[,1], PC2 = pc$x[,2], prop_mapped = prop_mapped) %>%
  ggplot(aes(PC1, PC2, color = prop_mapped)) +
  scale_colour_gradientn(colours = terrain.colors(10)) +
  geom_point(alpha = 0.5) +
  theme_bw()

data.frame(PC2 = pc$x[,2], PC3 = pc$x[,3], prop_mapped = prop_mapped) %>%
  ggplot(aes(PC2, PC3, color = prop_mapped)) +
  scale_colour_gradientn(colours = terrain.colors(10)) +
  geom_point(alpha = 0.5) +
  theme_bw()

# By patient
data.frame(PC1 = pc$x[,1], PC2 = pc$x[,2], patient = tumor_labels) %>%
  ggplot(aes(PC1, PC2, color = patient)) +
  geom_point(alpha = 0.5) +
  theme_bw()

data.frame(PC1 = pc$x[,1], PC3 = pc$x[,3], patient = tumor_labels) %>%
  ggplot(aes(PC1, PC3, color = patient)) +
  geom_point(alpha = 0.5) +
  theme_bw()
```

```{r, fig.height = 5, fig.width = 5}
# K-means clustering
set.seed(1)

# Elbow plot with SSE
sse = c()
for(k in 1:10){
  clust.kmeans <- kmeans(reducedDim(sce_bygene_transcripts, "PCA"), centers = k)
  sse = c(sse, clust.kmeans$tot.withinss)
}
plot(sse)

# Plot k-means
clust.kmeans <- kmeans(reducedDim(sce_bygene_transcripts, "PCA"), centers = 6)
colLabels(sce_bygene_transcripts) <- factor(clust.kmeans$cluster)
plotReducedDim(sce_bygene_transcripts, "PCA", colour_by="label")
```


## [Unrevised] GLM-PCA

### pre-mRNA

```{r}
sce = sce_prem
dat = dat_prem
```

```{r eval = F}
X = dat

# Takes a long time, may want to try on cluster
# Time increases with number of rows and number of latent dimensions ("L")
tic("glm PCA")
glmpc = glmpca(X, L = 3, fam = c("poi"))
toc()
```

```{r eval = F}
tumor_labels = gsub("-.*$", "", colnames(sce))

# Plot PC
data.frame(PC1 = glmpc$factors[,1], PC2 = glmpc$factors[,2], color = tumor_labels) %>%
  ggplot(aes(PC1, PC2, color = color)) +
  geom_point(alpha = 0.5) +
  theme_bw()

data.frame(PC1 = glmpc$factors[,1], PC3 = glmpc$factors[,3], color = tumor_labels) %>%
  ggplot(aes(PC1, PC3, color = color)) +
  geom_point(alpha = 0.5) +
  theme_bw()

data.frame(PC2 = glmpc$factors[,2], PC3 = glmpc$factors[,3], color = tumor_labels) %>%
  ggplot(aes(PC2, PC3, color = color)) +
  geom_point(alpha = 0.5) +
  theme_bw()
```

## Clustering

https://osca.bioconductor.org/clustering.html#k-means-clustering

```{r}

```

