---
title: qc and analysis of single nuc-seq data (alsf-filbin)
author: Stephanie Hicks and Albert Kuo
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{bash}
# Commands to start an interactive session on the JHPCE cluster
qrsh -l mem_free=10G,h_vmem=10G
module load conda_R
cd /fastscratch/myscratch/akuo/alsf-filbin
R
```

```{r}
library(here)
```

# Creating SummarizedExperiment object

First, we create a table with information about where each file (quantified counts from salmon) is. 
This will be used to create a `SummarizedExperiment` object.

```{r}
# list tumor names
tumor_names <- list.files(here("sample_data"))[
                  !grepl("*.txt", list.files(here("sample_data")))]

unique_sf_paths <- NULL
for(tum in tumor_names){
  ids <- list.files(here("sample_data", tum))
  ids <- unique(stringr::str_sub(ids, end=-13))
  ids <- here("salmon_quants", paste0(ids, "_quant"), "quant.sf")
  unique_sf_paths <- c(unique_sf_paths, ids)
}

unique_sf_ids <- NULL
for(tum in tumor_names){
  ids <- list.files(here("sample_data", tum))
  ids <- unique(stringr::str_sub(ids, end=-13))
  unique_sf_ids <- c(unique_sf_ids, ids)
}

coldata <- data.frame(files = unique_sf_paths, names = unique_sf_ids)
```

The only way I've figured out how to make this work is with `skipMeta = TRUE`. We need to figure out how to use the linkedTxome correctly so that `rowRanges(se)` isn't `NULL`. Then `tximeta` would be able to match the transcripts to genes for us.

```{r}
suppressPackageStartupMessages({
  library(tximeta)
})

se_file_name = here("salmon_quants", "se.rds")

if(!file.exists(se_file_name)){
  # Create SummarizedExperiment object
  se <- tximeta(coldata, skipMeta = TRUE) # Takes a few minutes, file size = 597 MB
  saveRDS(se, se_file_name)
} else {
  se = readRDS(se_file_name) # Takes a couple of seconds
}
```

## Explore SummarizedExperiment object

```{r}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(DESeq2)
})

colData(se)
assayNames(se)
rowRanges(se)
```

```{r}
dat <- assay(se, "counts")
dat <- dat[!(rowSums(dat) == 0),]
dim(dat)
dat[grep("ENST00000355537.4", rownames(dat)),]
summary(colSums(dat))
hist(colSums(dat))
```

```{r}
# percentages
premrna_prop = colSums(dat[grepl(".premrna", rownames(dat)),]) / colSums(dat)
mrna_prop = colSums(dat[!grepl(".premrna", rownames(dat)),]) / colSums(dat) # same as 1 - premrna_prop
summary(premrna_prop)
```

```{r}
## qqplot for a poisson
plot(quantile(rpois(1000, lambda = mean(dat[,1])),probs = seq(0.01,0.99,0.01)), 
     quantile(dat[,1],probs = seq(0.01,0.99,0.01)))
```

## Create SingleCellExperiment object

A SingleCellExperiment class is derived from the SummarizedExperiment class. The most important change is the addition of a new slot called `reducedDims`. Read more [here](https://osca.bioconductor.org/data-infrastructure.html#a-brief-recap-from-se-to-sce).

```{r}
# BiocManager::install('SingleCellExperiment')
library(SingleCellExperiment)
```

```{r}
sce = SingleCellExperiment(assays = list(counts = dat))

# you can access counts by assay(sce, "counts") or counts(sce)
# you can add a new entry to assays slot by assay(sce, "counts_new") = dat_new
```

## Quality Control

There are a couple of [QC metrics](https://osca.bioconductor.org/quality-control.html#choice-of-qc-metrics) to identify low-quality cells:

1. Using counts, i.e. cells with (a) a small library size (total sum of counts) `low_lib_size` or (b) few expressed endogeneous genes (nonzero counts for those genes) `low_n_features`
2. Using "spike-in transcripts", i.e. any enrichment of spike-in transcripts (higher proportion) 
3. Using the mitochondrial genome, i.e. any enrichment of reads in the mitochondrial genome is indicative of loss of cytoplasmic RNA

I will only do (1) for now.

```{r}
library(scater)

# Compute quality control metrics:
# sum is the total count for each cell
# detected contains the number of detected genes (actually transcripts for our data)
df = perCellQCMetrics(sce)
df
```

```{r}
# Find outliers with low library sizes (LibSize) and few detected features (n_features)
reasons = quickPerCellQC(df) # DataFrame of logical values
colSums(as.matrix(reasons))
```

```{r}
# Discard outliers
filtered = sce[, !reasons$discard] # 42 out of 576 cells were discarded
filtered
```

Diagnostic plots: https://osca.bioconductor.org/quality-control.html#quality-control-plots 
