---
title: "Generate figures"
author: "Albert Kuo"
date: "9/23/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages}
library(pacman)
p_load(cowplot, here, stringr, tidyverse, xtable, janitor)
```


# Figure 1 (distributions)

```{r}
a = readRDS(here("./mouse_cortex/plots/meanvar_plot_preandmrna_excitatory_neuron_cortex_1.rds"))
b = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_excitatory_neuron_cortex_1.rds"))
c = readRDS(here("./mouse_cortex/plots/meanvar_plot_preandmrna_inhibitory_neuron_cortex_1.rds"))
d = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_inhibitory_neuron_cortex_1.rds"))
e = readRDS(here("./mouse_cortex/plots/meanvar_plot_preandmrna_astrocyte_cortex_1.rds"))
f = readRDS(here("./mouse_cortex/plots/prob0_plot_preandmrna_astrocyte_cortex_1.rds"))

# Edit titles and labels
a = a +
  labs(title = "Excitatory neuron",
       subtitle = NULL) +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6),
        axis.title = element_text(size = 8),
        plot.title = element_text(size = 12))
b = b + 
  labs(title = "Excitatory neuron",
       subtitle = NULL) +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6),
        axis.title = element_text(size = 8),
        plot.title = element_text(size = 12))
c = c +
  labs(title = "Inhibitory neuron",
       subtitle = NULL) +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6),
        axis.title = element_text(size = 8),
        plot.title = element_text(size = 12))
d = d + 
  labs(title = "Inhibitory neuron",
       subtitle = NULL) +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6),
        axis.title = element_text(size = 8),
        plot.title = element_text(size = 12))
e = e +
  labs(title = "Astrocyte",
       subtitle = NULL) +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6),
        axis.title = element_text(size = 8),
        plot.title = element_text(size = 12))
f = f + 
  labs(title = "Astrocyte",
       subtitle = NULL) +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6),
        axis.title = element_text(size = 8),
        plot.title = element_text(size = 12))


p = plot_grid(a, b, c, d, e, f, nrow = 3, labels = "auto")
print(p)
ggsave(here("./publication/figures/figure_1.png"), plot = p, width = 12, height = 6)
```

# Figure 3 (cell type misclassification)

```{r}
celltypemis = readRDS(here("./mouse_cortex/plots/celltypemis_plot.rds"))
celltypemis = celltypemis +
  theme(axis.text.x = element_text(size = 7))
astrocytemgratios = readRDS(here("./mouse_cortex/plots/astrocytemgratios_plot.rds"))

p = plot_grid(celltypemis, astrocytemgratios, nrow = 1, rel_widths = c(0.7, 0.3), labels = "auto")
print(p)
ggsave(here("./publication/figures/figure_3.png"), plot = p, width = 16, height = 8)
```

# Supplementary Figure 1 (lib size by gene type)

```{r}
a = readRDS(here("./mouse_cortex/plots/lib_size_protein_coding_comparison.rds")) 
b = readRDS(here("./mouse_cortex/plots/lib_size_lincRNA_comparison.rds"))
c = readRDS(here("./mouse_cortex/plots/lib_size_antisense_comparison.rds"))
d = readRDS(here("./mouse_cortex/plots/lib_size_processed_pseudogene_comparison.rds"))

a = a +
  labs(title = "Protein coding",
       y = "Total reads")

b = b +
  labs(title = "Long non-coding RNA",
       y = "Total reads")

c = c +
  labs(title = "Antisense",
       y = "Total reads")

d = d +
  labs(title = "Processed pseudogene",
       y = "Total reads")

p = plot_grid(a, b, c, d, nrow = 2, labels = "auto")
print(p)
ggsave(here("./publication/figures/sfigure_1.png"), plot = p, width = 16, height = 8)
```


# Supplementary Figure 2 (PCA by cell type)

```{r}
a = readRDS(here("./mouse_cortex/plots/pcading_plot_transcripts.rds"))
b = readRDS(here("./mouse_cortex/plots/pcading_plot_preandmrna.rds"))
c = readRDS(here("./mouse_cortex/plots/pcading_plot_introncollapse.rds"))
d = readRDS(here("./mouse_cortex/plots/pcading_plot_intronseparate.rds"))

a = a + labs(title = "Transcripts pipeline")
b = b + labs(title = "Preandmrna pipeline")
c = c + labs(title = "Introncollapse pipeline")
d = d + labs(title = "Intronseparate pipeline")

p = plot_grid(a, b, c, d, nrow = 2, labels = "auto")
print(p)
ggsave(here("./publication/figures/sfigure_2.png"), plot = p, width = 12, height = 8)
```


# Supplementary Figure 3 (PCA by cortex)

```{r}
a = readRDS(here("./mouse_cortex/plots/pcacortex_plot_transcripts.rds"))
b = readRDS(here("./mouse_cortex/plots/pcacortex_plot_preandmrna.rds"))
c = readRDS(here("./mouse_cortex/plots/pcacortex_plot_introncollapse.rds"))
d = readRDS(here("./mouse_cortex/plots/pcacortex_plot_intronseparate.rds"))

a = a + labs(title = "Transcripts pipeline")
b = b + labs(title = "Preandmrna pipeline")
c = c + labs(title = "Introncollapse pipeline")
d = d + labs(title = "Intronseparate pipeline")

p = plot_grid(a, b, c, d, nrow = 2, labels = "auto")
print(p)
ggsave(here("./publication/figures/sfigure_3.png"), plot = p, width = 12, height = 8)
```

# Supplementary Table (BIC values)

```{r}
files_ls = list.files(here("./mouse_cortex/plots"), pattern = "bic_data.*.rds", full.names = T)
bic_values = lapply(files_ls, readRDS)
for(i in seq_along(bic_values)){
  x = bic_values[i]
  cortex = str_sub(files_ls[i], -12, -5)
  x[[1]]$cortex = cortex
  x[[1]]$pipeline = str_split(files_ls[i], "_")[[1]][4]
  if("multinomial" %in% names(x[[1]])){
    print(i)
    x[[1]] = x[[1]] %>%
      dplyr::rename(binomial = multinomial)
  }
  x[[1]] = x[[1]] %>%
    select(pipeline, cortex, cell_type_name, binomial, poisson, nbinomial, nbinomial_2)
  bic_values[i] = x
}

bic_values = bind_rows(bic_values) %>%
  arrange(pipeline, cortex, cell_type_name)
names(bic_values) = c("Pipeline", "Cortex", "Cell type", "Binomial BIC", "Poisson BIC", "Negative binomial BIC", "Negative binomial 2 BIC")

# Print out which distribution has lowest BIC
bic_values %>% 
  pivot_longer(cols = ends_with("BIC")) %>%
  group_by(Pipeline, Cortex, `Cell type`) %>%
  filter(value == min(value)) %>%
  ungroup() %>%
  tabyl(name)

print(xtable(bic_values, digits = 0, label = "table:bic", caption = "BIC log-likelihoods for every pipeline, cell type, and cortex combination. Negative binomial refers to a negative binomial distribution with one overdispersion parameter for all genes. Negative binomial 2 refers to a negative binomial distribution with gene-specific overdispersion parameters."), include.rownames = F)
```

