---
title: "droplet single-cell RNA data"
author: "Albert Kuo"
date: "1/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Investigate distributions in single-cell RNA data:

* https://www.nature.com/articles/s41587-019-0379-5 
* https://www.nxn.se/valent/2017/11/16/droplet-scrna-seq-is-not-zero-inflated
* https://data.caltech.edu/records/1264

How to read in H5AD files into R:
* http://www.bioconductor.org/packages/devel/bioc/vignettes/zellkonverter/inst/doc/zellkonverter.html

We want to apply goodness-of-fit tests to the negative control single-cell RNA data used by Valentine and see how it compares to results for our single-nucleus RNA mouse cortex data.

## Data

```{r}
suppressPackageStartupMessages({
  library(zellkonverter)
  library(here)
  library(SingleCellExperiment)
  library(tictoc)
  library(tidyverse)
})
source(here("./mouse_cortex/code/distribution-plots-helpers.R"))
```

```{r}
svensson_1 = readH5AD(here("./scrna/Svensson et al 2017 (1).h5ad"))
svensson_2 = readH5AD(here("./scrna/Svensson et al 2017 (2).h5ad"))
zheng = readH5AD(here("./scrna/Zheng et al 2017.h5ad"))
klein = readH5AD(here("./scrna/Klein et al 2015.h5ad"))
```

```{r}
h5ad_ls = list(svensson_1, svensson_2, zheng, klein)
```


```{r}
set.seed(1)
counts_ls = list()
for(i in seq_along(h5ad_ls)){
  counts = assay(h5ad_ls[[i]], "X")
  
  tic("downsample")
  counts_scaled = Down_Sample_Matrix(ceiling(counts))
  toc()
  
  counts_scaled = counts_scaled[rowSums(counts_scaled) != 0, ]
  counts_ls[[i]] = counts_scaled
}
```

```{r}
# Simulated datasets with same dimensions as snRNA data
emp_values = readRDS(here("./mouse_cortex/output/emp_values_edger.rds"))
means = emp_values$means
sizes = emp_values$sizes
counts_ls[[i+1]] = matrix(rpois(n = 347*21483, lambda = 0.1), ncol = 347, nrow = 21483) # poisson
counts_ls[[i+2]] = matrix(rpois(n = 347*21483, lambda = 0.3), ncol = 347, nrow = 21483) # poisson 
counts_ls[[i+3]] = matrix(rpois(n = 347*21483, lambda = 0.5), ncol = 347, nrow = 21483) # poisson 
counts_ls[[i+4]] = matrix(rpois(n = 347*21483, lambda = 0.7), ncol = 347, nrow = 21483) # poisson 
counts_ls[[i+5]] = matrix(rpois(n = 347*21483, lambda = 1), ncol = 347, nrow = 21483) # poisson 
counts_ls[[i+6]] = t(sapply(seq_len(length(means)), function(i) 
  rpois(n = 347, lambda = means[i]))) # poisson with different parameters for each gene
counts_ls[[i+7]] = matrix(rnbinom(n = 347*21483, size = 63, mu = 0.1), ncol = 347, nrow  = 21483) # negative binomial
counts_ls[[i+8]] = matrix(rnbinom(n = 347*21483, size = 63, mu = 0.3), ncol = 347, nrow  = 21483) # negative binomial
counts_ls[[i+9]] = matrix(rnbinom(n = 347*21483, size = 63, mu = 0.5), ncol = 347, nrow  = 21483) # negative binomial
counts_ls[[i+10]] = matrix(rnbinom(n = 347*21483, size = 63, mu = 0.7), ncol = 347, nrow  = 21483) # negative binomial
counts_ls[[i+11]] = matrix(rnbinom(n = 347*21483, size = 63, mu = 1), ncol = 347, nrow  = 21483) # negative binomial
counts_ls[[i+12]] = t(sapply(seq_len(length(means)), function(k) 
  rnbinom(n = 347, size = sizes[k], mu = means[k]))) # negative binomial with different parameters for each gene

for(j in (i+1):length(counts_ls)){
  tic("downsample")
  counts_ls[[j]] = Down_Sample_Matrix(ceiling(counts_ls[[j]]))
  toc()
}

saveRDS(counts_ls, here("./scrna/counts_ls.rds"))
```

```{r}
counts_ls = readRDS(here("./scrna/counts_ls.rds"))
```


#### Goodness-of-fit tests

##### By Count Values

```{r}
plot_dt_pois_ls = list()
plot_dt_nb_ls = list()
for(i in seq_along(counts_ls)){
  counts = counts_ls[[i]]
  
  # Poisson
  p_values_pois = p_chisq_test(counts, distribution = "poisson")
  plot_dt_pois_ls[[i]] = tibble(p_value = p_values_pois[[1]],
                             expression = rowSums(counts),
                             i = i)
  
  # Negative binomial
  p_values_nb = p_chisq_test(counts, distribution = "nb")
  plot_dt_nb_ls[[i]] = tibble(p_value = p_values_nb[[1]],
                           expression = rowSums(counts)[setdiff(1:nrow(counts), p_values_nb[[2]])],
                           i = i)
}
```

```{r}
# Poisson plots
plot_dt_pois = bind_rows(plot_dt_pois_ls)
plot_dt_pois %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value)) + 
  facet_wrap(~ i, scales = "free") +
  geom_histogram() + 
  theme_bw()

plot_dt_pois %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value, y = expression)) + 
  facet_wrap(~ i, scales = "free") +
  geom_point(size = 0.1, alpha = 0.1) +
  geom_smooth() +
  scale_y_log10() +
  theme_bw()
```


```{r}
# Negative binomial plots
plot_dt_nb = bind_rows(plot_dt_nb_ls)
plot_dt_nb %>%
  filter(expression > 50) %>%
  ggplot(aes(x = p_value)) + 
  facet_wrap(~ i, scales = "free") +
  geom_histogram() + 
  theme_bw()

plot_dt_nb %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value, y = expression)) + 
  facet_wrap(~ i, scales = "free") +
  geom_point(size = 0.1, alpha = 0.1) +
  geom_smooth() +
  scale_y_log10() +
  theme_bw()
```

##### By Sample

This is based on Figure 2 from this paper: https://genome.cshlp.org/content/18/9/1509.full.html 

```{r}
plot_dt_pois_ls = list()
plot_dt_nb_1_ls = list()
plot_dt_nb_2_ls = list()
plot_dt_nb_3_ls = list()
for(i in seq_along(counts_ls)){
  print(i)
  counts = counts_ls[[i]]
  
  # Poisson
  chi_obs_pois = p_chisq_test_2(counts, distribution = "poisson")
  plot_dt_pois_ls[[i]] = tibble(chi_obs = chi_obs_pois,
                                expression = rowSums(counts),
                                var = apply(counts, 1, var),
                                prop_0 = apply(counts, 1, function(x) sum(x == 0))) %>%
    mutate(id = 1:n()) %>%
    arrange(chi_obs) %>%
    drop_na() %>%
    mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = ncol(counts) - 1),
           percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
                                        chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
                                        chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
                                        T ~ "rest"),
           i = i)
  
  # Negative binomial (single overdispersion parameter)
  chi_obs_nb = p_chisq_test_2(counts, distribution = "nb 1")
  plot_dt_nb_1_ls[[i]] = tibble(chi_obs = chi_obs_nb,
                                expression = rowSums(counts),
                                var = apply(counts, 1, var),
                                prop_0 = apply(counts, 1, function(x) sum(x == 0))) %>%
    mutate(id = 1:n()) %>%
    arrange(chi_obs) %>%
    drop_na() %>%
    mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = ncol(counts) - 2),
           percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
                                        chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
                                        chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
                                        T ~ "rest"),
           i = i)
  
  # Negative binomial (gene-specific overdispersion parameter)
  chi_obs_nb = p_chisq_test_2(counts, distribution = "nb 2")
  plot_dt_nb_2_ls[[i]] = tibble(chi_obs = chi_obs_nb[[1]],
                                f_phi = chi_obs_nb[[2]],
                                expression = rowSums(counts),
                                var = apply(counts, 1, var),
                                prop_0 = apply(counts, 1, function(x) sum(x == 0))) %>%
    mutate(id = 1:n()) %>%
    arrange(chi_obs) %>%
    drop_na() %>%
    mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = ncol(counts) - 2),
           percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
                                        chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
                                        chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
                                        T ~ "rest"),
           i = i)
  
  # Negative binomial with true size parameter
  m = counts
  total_sum = sum(m)
  lambda_j = rowSums(m)/total_sum # proportion of counts in gene j
  c_i = colSums(m)/total_sum # proportion of counts in sample i (should be approximately same across i for downsampled matrix)
  mu_ij = outer(lambda_j, c_i)*total_sum
  
  f_obs = m
  f_hyp = mu_ij
  f_var = mu_ij + mu_ij^2/sizes
  
  chi_obs_nb = rowSums((f_obs-f_hyp)^2/f_var)
  plot_dt_nb_3_ls[[i]] = tibble(chi_obs = chi_obs_nb,
                                expression = rowSums(counts),
                                var = apply(counts, 1, var),
                                prop_0 = apply(counts, 1, function(x) sum(x == 0))) %>%
    mutate(id = 1:n()) %>%
    arrange(chi_obs) %>%
    drop_na() %>%
    mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = ncol(counts) - 2),
           percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
                                        chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
                                        chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
                                        T ~ "rest"),
           i = i)
}
```


```{r}
# Poisson plots
plot_dt_pois = bind_rows(plot_dt_pois_ls)
plot_dt_pois %>%
  # filter(expression > 50) %>%
    ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free") +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```
```{r}
plot_dt_pois %>%
  ggplot(aes(x = expression, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_pois %>%
  ggplot(aes(x = prop_0, color = percentile_color)) +
  geom_density() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_pois %>%
  ggplot(aes(x = var, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
```

```{r}
# Negative binomial plots (single overdispersion parameter)
plot_dt_nb_1 = bind_rows(plot_dt_nb_1_ls)
plot_dt_nb_1 %>%
  # filter(expression > 50) %>%
    ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free") +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```

```{r}
plot_dt_nb_1 %>%
  ggplot(aes(x = expression, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_nb_1 %>%
  ggplot(aes(x = prop_0, color = percentile_color)) +
  geom_density() +
  # scale_x_log10() + 
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_nb_1 %>%
  ggplot(aes(x = var, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
```

```{r}
# Negative binomial plots (gene-specific overdispersion parameter)
plot_dt_nb_2 = bind_rows(plot_dt_nb_2_ls)
plot_dt_nb_2 %>%
  # filter(expression > 50) %>%
    ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free") +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```

```{r}
plot_dt_nb_3_ls[[1]] %>% ggplot(aes(x = chi_quantile, y = chi_obs, color = percentile_color)) + geom_point(size = 0.5) + geom_abline(slope = 1, intercept = 0) + theme_bw()
```

```{r}
plot_dt_nb_2 %>%
  ggplot(aes(x = expression, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_nb_2 %>%
  ggplot(aes(x = prop_0, color = percentile_color)) +
  geom_density() +
  # scale_x_log10() + 
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
plot_dt_nb_2 %>%
  ggplot(aes(x = var, color = percentile_color)) +
  geom_density() +
  scale_x_log10() +
  facet_wrap(~ i, ncol = 1, scales = "free_y") +
  theme_bw()
```

Diagnosing parameter estimates

```{r}
emp_values = readRDS(here("./mouse_cortex/output/emp_values_edger.rds"))
means = emp_values$means
sizes = emp_values$sizes
params = tibble(means = means,
                sizes = sizes,
                id = 1:length(means))
```


```{r}
plot_dt_nb_2 = plot_dt_nb_2 %>%
  left_join(., params, by = "id")

# Mean vs size (emp parameters)
plot_dt_nb_2 %>%
  ggplot(aes(x = means, y = sizes)) +
  geom_point(alpha = 0.01) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()

# Mean vs phi
plot_dt_nb_2 %>%
  ggplot(aes(x = means, y = f_phi)) +
  geom_point(alpha = 0.01) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()

# Size vs phi
plot_dt_nb_2 %>%
  # filter(!(sizes %in% common_sizes_1) & !(f_phi %in% common_sizes_2)) %>%
  ggplot(aes(x = sizes, y = f_phi)) +
  geom_point(alpha = 0.01) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()
```

```{r}
# Common size estimates (note that the overall phi estimate is 2.16)
common_sizes_1 = plot_dt_nb_2 %>% count(sizes) %>% filter(n > 1) %>% pull(sizes) # 0.339 and 10240 occurs in 6600 and 4000 genes
common_sizes_2 = plot_dt_nb_2 %>% count(f_phi) %>% filter(n > 1) %>% pull(f_phi) # 0.501 and 10240 occurs in 6600 and 5000 genes
```

```{r}
# What are the genes that get assigned these size estimates?
size_param_groups = tibble(size_label = c("size_0.339", "size_10240"),
                           sizes = common_sizes_1)
f_phi_param_groups = tibble(f_phi_label = c("f_phi_0.501", "f_phi_10240"),
                            f_phi = common_sizes_2)
                           
plot_dt_nb_2 = plot_dt_nb_2 %>%
  left_join(., size_param_groups, by = "sizes") %>%
  left_join(., f_phi_param_groups, by = "f_phi")
```

```{r}
plot_dt_nb_2 %>%
  ggplot(aes(x = prop_0, y = expression)) +
  geom_point(size = 0.1) +
  facet_wrap(~ size_label + f_phi_label) +
  scale_y_log10() +
  theme_bw()

plot_dt_nb_2 %>%
  ggplot(aes(x = var, y = expression)) +
  geom_point(size = 0.1) +
  facet_wrap(~ size_label + f_phi_label) +
  scale_y_log10() +
  scale_x_log10() +
  theme_bw()
```


```{r}
# Remove common size parameters
plot_dt_nb_2 %>%
  filter(!(sizes %in% common_sizes_1[2]) & !(f_phi %in% common_sizes_2[2])) %>%
  ggplot(aes(x = sizes, y = f_phi)) +
  geom_point(alpha = 0.01) +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()

plot_dt_nb_2_filter = plot_dt_nb_2 %>% 
  filter(!(sizes %in% common_sizes_1[2]) & !(f_phi %in% common_sizes_2[2])) %>%
  arrange(chi_obs) %>%
  mutate(chi_quantile = qchisq(p = (1:length(chi_obs))/length(chi_obs), df = ncol(counts) - 2),
         percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995, na.rm = TRUE) ~ "> 99.5",
                                      chi_obs > quantile(chi_obs, 0.95, na.rm = TRUE) ~ "> 95",
                                      chi_obs > quantile(chi_obs, 0.90, na.rm = TRUE) ~ "> 90",
                                      T ~ "rest"))
plot_dt_nb_2_filter %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  facet_wrap(~ i, scales = "free") +
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()
```


