---
title: "droplet single-cell RNA data"
author: "Albert Kuo"
date: "1/14/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Investigate distributions in single-cell RNA data:

* https://www.nature.com/articles/s41587-019-0379-5 
* https://www.nxn.se/valent/2017/11/16/droplet-scrna-seq-is-not-zero-inflated
* https://data.caltech.edu/records/1264

How to read in H5AD files into R:
* http://www.bioconductor.org/packages/devel/bioc/vignettes/zellkonverter/inst/doc/zellkonverter.html

We want to apply goodness-of-fit tests to the negative control single-cell RNA data used by Valentine and see how it compares to results for our single-nucleus RNA mouse cortex data.

## Data

```{r}
suppressPackageStartupMessages({
  library(zellkonverter)
  library(here)
  library(SingleCellExperiment)
  library(tictoc)
  library(tidyverse)
})
source(here("./mouse_cortex/code/distribution-plots-helpers.R"))
```

```{r}
svensson_1 = readH5AD(here("./scrna/Svensson et al 2017 (1).h5ad"))
svensson_2 = readH5AD(here("./scrna/Svensson et al 2017 (2).h5ad"))
zheng = readH5AD(here("./scrna/Zheng et al 2017.h5ad"))
klein = readH5AD(here("./scrna/Klein et al 2015.h5ad"))
```

```{r}
h5ad_ls = list(svensson_1, svensson_2, zheng, klein)
```


```{r}
set.seed(1)
counts_ls = list()
for(i in seq_along(h5ad_ls)){
  counts = assay(h5ad_ls[[i]], "X")
  
  tic("downsample")
  counts_scaled = Down_Sample_Matrix(ceiling(counts))
  toc()
  
  counts_scaled = counts_scaled[rowSums(counts_scaled) != 0, ]
  counts_ls[[i]] = counts_scaled
}
```

#### Goodness-of-fit tests

##### By Count Values

```{r eval = F}
plot_dt_pois_ls = list()
plot_dt_nb_ls = list()
for(i in seq_along(counts_ls)){
  counts = counts_ls[[i]]
  
  # Poisson
  p_values_pois = p_chisq_test(counts, distribution = "poisson")
  plot_dt_pois_ls[[i]] = tibble(p_value = p_values_pois[[1]],
                             expression = rowSums(counts),
                             i = i)
  
  # Negative binomial
  p_values_nb = p_chisq_test(counts, distribution = "nb")
  plot_dt_nb_ls[[i]] = tibble(p_value = p_values_nb[[1]],
                           expression = rowSums(counts)[setdiff(1:nrow(counts), p_values_nb[[2]])],
                           i = i)
}
```

```{r}
# Poisson plots
plot_dt_pois = bind_rows(plot_dt_pois_ls)
plot_dt_pois %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value)) + 
  facet_wrap(~ i) +
  geom_histogram() + 
  theme_bw()

plot_dt_pois %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value, y = expression)) + 
  facet_wrap(~ i) +
  geom_point(size = 0.1, alpha = 0.1) +
  geom_smooth() +
  scale_y_log10() +
  theme_bw()
```


```{r}
# Negative binomial plots
plot_dt_nb = bind_rows(plot_dt_nb_ls)
plot_dt_nb %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value)) + 
  facet_wrap(~ i) +
  geom_histogram() + 
  theme_bw()

plot_dt_nb %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value, y = expression)) + 
  facet_wrap(~ i) +
  geom_point(size = 0.1, alpha = 0.1) +
  geom_smooth() +
  scale_y_log10() +
  theme_bw()
```

##### By Sample

This is based on Figure 2 from this paper: https://genome.cshlp.org/content/18/9/1509.full.html 

```{r}
# Poisson
total_sum = sum(counts_sub_cluster_scaled)
lambda_j = rowSums(counts_sub_cluster_scaled)/total_sum # proportion of counts in gene j
c_i = colSums(counts_sub_cluster_scaled)/total_sum # proportion of counts in sample i (should be approximately same across i for downsampled matrix)
mu_ij = outer(lambda_j, c_i)*total_sum

f_obs = counts_sub_cluster_scaled
f_hyp = mu_ij

chiSquare.pois = rowSums((f_obs-f_hyp)^2/f_hyp) # chi-squared statistic for each gene
df = ncol(f_obs) - 1

plot_dt = tibble(chi_obs = sort(chiSquare.pois),
                 chi_quantile = qchisq(p = (1:length(chiSquare.pois))/length(chiSquare.pois),
                                       df = df),
                 percentile_color = case_when(chi_obs > quantile(chi_obs, 0.995) ~ "> 99.5",
                                        chi_obs > quantile(chi_obs, 0.95) ~ "> 95",
                                        chi_obs > quantile(chi_obs, 0.90) ~ "> 90",
                                        T ~ "rest"),
                 expression = rowSums(counts_sub_cluster_scaled))

plot_dt %>%
  ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()

plot_dt %>%
  filter(expression > 50) %>%
  ggplot(aes(x = chi_quantile,
             y = chi_obs,
             color = percentile_color)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1, intercept = 0) +
  # coord_cartesian(ylim = c(250, 1000)) +
  theme_bw()

# Histogram of p-values
plot_dt = plot_dt %>%
  mutate(p_value = 1 - pchisq(chi_obs, df = ncol(counts_sub_cluster_scaled) - 1))

plot_dt %>%
  # filter(expression > 50) %>%
  ggplot(aes(x = p_value)) + 
  geom_histogram() +
  theme_bw()
```
