---
title: "single-nucleus analysis (mouse cortex)"
author: "Albert Kuo and Stephanie Hicks"
date: "9/8/2020"
output: 
  html_document:
    code_folding: "hide"
    toc: TRUE
    toc_float: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Exploratory Analysis

```{r}
suppressPackageStartupMessages({
  library(here)
  library(tidyverse)
  library(tictoc)
  library(scran)
  library(scater)
  library(BiocSingular)
  library(SingleCellExperiment)
})
```

```{r}
run_number = "all" # give run_number or "all" for all of them together
sce_ls = list()
sce_ls[["transcripts"]] = readRDS(here("mouse_cortex", "salmon_quants", "transcripts_pipeline", paste0("sce_", run_number, ".rds")))
sce_ls[["preandmrna"]] = readRDS(here("mouse_cortex", "salmon_quants", "preandmrna_pipeline", paste0("sce_", run_number, ".rds")))
sce_ls[["introncollapse"]] = readRDS(here("mouse_cortex", "salmon_quants", "introncollapse_pipeline", paste0("sce_", run_number, ".rds")))
sce_ls[["intronseparate"]] = readRDS(here("mouse_cortex", "salmon_quants", "intronseparate_pipeline", paste0("sce_", run_number, ".rds")))
```

### Quality control

Discard low-quality cells and convert from SummarizedExperiment to SingleCellExperiment. Done in `save-sce.R`.

### Library size comparison

To generate and save these plots, run `libsize-plots.R`.

```{r}
# Get library sizes
lib_sizes_ls = list()
for(i in seq_along(sce_ls)){
  sce = sce_ls[[i]]
  pipeline = names(sce_ls)[i]
  
  lib_sizes_ls[[pipeline]] = colData(sce) %>%
    as_tibble() %>%
    mutate(cell_barcode = rownames(colData(sce)),
           pipeline = pipeline)
}

lib_sizes = bind_rows(lib_sizes_ls)

# Compute median label
summ = lib_sizes %>% 
  group_by(pipeline, cortex) %>% 
  summarize(median = median(lib_size))

# Plot library size comparison between pipelines
lib_sizes %>%
  ggplot(aes(x = pipeline, y = lib_size)) +
  geom_boxplot() + 
  # geom_text(data = summ, aes(x = pipeline, y = median, 
                              # label = round(median, 0)), 
            # vjust = -0.5, size = 2.5) +
  facet_wrap(~ cortex) +
  labs(x = "Pipeline",
       y = "Library size") +
  theme_bw() +
  theme(axis.text = element_text(size = 8))

ggsave(file = here(paste0("./mouse_cortex/plots/lib_size_comparison.png")),
       width = 8, height = 5)
```

### Mapping differences by gene

```{r}
# Calculate sums for different gene types
gene_sum_tb_ls = list()
for(i in seq_along(sce_ls)){
  pipeline_name = names(sce_ls)[i]
  sce = sce_ls[[i]]
  
  lib_size_tb = colData(sce) %>%
    as_tibble() %>%
    mutate(cell_barcode = colnames(sce))
  
  for(gene_type_name in unique(rowData(sce)$gene_type)){ 
    genes_in_type = rowData(sce)[rowData(sce)$gene_type == gene_type_name, "gene_id"]
    m = counts(sce)[genes_in_type, ]
    if(length(genes_in_type) > 1){
      lib_size_gene_type = colSums(m)
    } else {
      lib_size_gene_type = m
    }
    lib_size_tb = lib_size_tb %>% 
      mutate(!!gene_type_name := lib_size_gene_type)
  }
  gene_sum_tb_ls[[pipeline_name]] = lib_size_tb %>%
    mutate(pipeline = pipeline_name)
}
gene_sum_tb = bind_rows(gene_sum_tb_ls)
```

```{r}
# Plot
for(gene_type in c("protein_coding", "lincRNA", "antisense", "processed_pseudogene")){ # 4 with highest counts across all pipelines
  # Compute median label
  summ = gene_sum_tb %>%
    group_by(pipeline, cortex) %>%
    summarize(median = median(!!sym(gene_type)))
  
  p = gene_sum_tb %>%
    ggplot(aes(x = pipeline, y = !!sym(gene_type))) + 
    geom_boxplot() +
    # geom_label(data = summ, aes(x = pipeline, y = median, 
    #                             label = round(median, 0))) +
    labs(x = "Pipeline",
         y = "Library size") +
    facet_wrap(~ cortex) +
    theme_bw()
  
  print(p)
}
```

```{r}
# pipeline = "preandmrna"
# sce = sce_ls[[pipeline]]
```

### PCA

PCA done in `save-sce.R`. To generate and save these plots, run `distribution-plots.R`.

```{r}
# PCA colored by provided attribute
plotReducedDim(sce_ls[["transcripts"]], "PCA", colour_by = "ding_labels", point_size = 1)
```


### Cell types 

Cell types added in `save-sce.R`.

#### singleR

Use singleR to map cells to cell type based on correlation in expression to mouse bulk RNA-seq data.

Vignettes:

* https://www.bioconductor.org/packages/release/bioc/vignettes/SingleR/inst/doc/SingleR.html
* https://bioconductor.org/packages/devel/data/experiment/vignettes/celldex/inst/doc/userguide.html#23_Mouse_RNA-seq

```{r}
library(pheatmap)

pred_celltypes = readRDS(here("mouse_cortex", "salmon_quants", "transcripts_pipeline", "singler_results.rds"))
all.markers <- metadata(pred_celltypes)$de.genes
# Built-in plots
# plotScoreHeatmap(pred_celltypes)
# features = unique(unlist(all.markers$qNSCs))
# features = tx2gene %>% filter(gene_name %in% features) %>% pull(gene_id) %>% unique()
# plotHeatmap(sce[, which(colData(sce)$ding_labels == "Astrocyte")], # Need to run ding section first
#             order_columns_by = "singleR_labels_pruned",
#             features = features,
#             fontsize_row = 1) # takes a while to run if too many cells/genes
```

```{r}
# PCA colored by singleR labels
plotReducedDim(sce, "PCA", colour_by = "singleR_labels_pruned")
```


#### Ding

> We used marker genes for each cell type to compute a score for each cell and automatically assign cell types to clusters.

Cell types added in `save-sce.R`.

```{r}
# PCA colored by Ding's labels
plotReducedDim(sce_ls[["preandmrna"]], "PCA", colour_by = "ding_labels")
```

#### Differences between pipelines

Some of these plots can be generated and saved by running `celltype-plots.R`. 

```{r}
# Get overlap with singleR labels
tb = as_tibble(colData(sce_ls[["transcripts"]]))
overlap = table(tb$singleR_labels_pruned, tb$ding_labels, useNA = "ifany") %>%
  as.data.frame()

# Normalize frequency within each Ding cell type
overlap = overlap %>%
  group_by(Var2) %>%
  mutate(Freq_norm = Freq/sum(Freq)) %>%
  ungroup()

# Plot
overlap %>%
  ggplot(aes(x = Var1, y = Var2, fill = Freq)) + 
  scale_fill_gradient(low="white", high="blue") +
  geom_tile() +
  labs(x = "singleR_labels_pruned",
       y = "ding_labels") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

overlap %>%
  ggplot(aes(x = Var1, y = Var2, fill = Freq_norm)) + 
  scale_fill_gradient(low="white", high="blue") +
  geom_tile() +
  labs(x = "singleR_labels_pruned",
       y = "ding_labels") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

```{r}
tb = as_tibble(preandmrna_cd) %>%
  mutate(cell = rownames(preandmrna_cd))

# Label cells where singleR label != ding label
tb = tb %>%
  mutate(ding_labels_match = case_when(ding_labels == "Astrocyte" ~ "Astrocytes",
                                       ding_labels == "Endothelial" ~ "Endothelial cells",
                                       ding_labels == "Excitatory neuron" ~ "Neurons",
                                       ding_labels == "Inhibitory neuron" ~ "Neurons",
                                       ding_labels == "Oligodendrocyte" ~ "Oligodendrocytes",
                                       T ~ ding_labels))

tb = tb %>%
  mutate(singleR_labels_match = case_when(singleR_labels_pruned == "Astrocytes activated" ~ "Astrocytes",
                                          singleR_labels_pruned == "Fibroblasts activated" ~ "Fibroblasts",
                                          singleR_labels_pruned == "Fibroblasts senescent" ~ "Fibroblasts",
                                          singleR_labels_pruned == "Macrophages activated" ~ "Macrophages",
                                          singleR_labels_pruned == "Microglia activated" ~ "Microglia",
                                          singleR_labels_pruned == "Neurons activated" ~ "Neurons",
                                          T ~ singleR_labels_pruned),
         non_matching_cells = (singleR_labels_match != ding_labels_match) | (is.na(singleR_labels_match) & !is.na(ding_labels_match)))

tb_misclassify = tb %>%
  filter(!is.na(ding_labels_match)) 

tb_misclassify %>%
  ggplot(aes(x = lib_size)) +
  geom_histogram(aes(fill = non_matching_cells)) +
  facet_wrap(~ ding_labels_match, nrow = 2, scales = "free") +
  labs(title = "preandmrna pipeline") +
  theme_bw()

print(sum(tb_misclassify$non_matching_cells, na.rm = T))

tb_misclassify
```

```{r}
# Counts for marker genes
astrocyte_mg = unique(unlist(all.markers$Astrocytes))
genes_in_type = t2g %>% filter(gene_name %in% astrocyte_mg) %>% pull(gene_id)
cd = colData(sce) %>% as_tibble() %>% mutate(cell = rownames(colData(sce)))
astrocyte_cells = cd %>% filter(ding_labels == "Astrocyte") %>% pull(cell)
astrocyte_matrix = counts(sce)[genes_in_type, astrocyte_cells] # Astrocyte cells and astrocyte marker genes only
# rowSums(astrocyte_matrix)
colSums(astrocyte_matrix) %>% 
  as_tibble() %>%
  mutate(cell = colnames(astrocyte_matrix)) %>%
  left_join(., cd, by = "cell") %>%
  ggplot(aes(x = value)) +
  geom_histogram(aes(fill = singleR_labels_pruned)) +
  theme_bw()
```

```{r}
# genes_in_type = names(sort(-rowSums(astrocyte_matrix))[3]) # Choose a specific gene
gene_type_name = "Astrocyte"
gene_sum_tb_ls = list()
for(i in seq_along(sce_ls)){
  pipeline_name = names(sce_ls)[i]
  sce_i = sce_ls[[i]]
  
  lib_size_tb = tibble(cell_barcode = astrocyte_cells)
  lib_size_gene_type = colSums(counts(sce_i)[intersect(rownames(counts(sce_i)), genes_in_type), astrocyte_cells, drop = F])
  lib_size_gene_type_tb = tibble(cell_barcode = names(lib_size_gene_type),
                                 !!gene_type_name := lib_size_gene_type)
  lib_size_tb = lib_size_tb %>% 
    left_join(., lib_size_gene_type_tb, by = "cell_barcode")
  gene_sum_tb_ls[[pipeline_name]] = lib_size_tb %>%
    pivot_longer(cols = -cell_barcode) %>%
    mutate(pipeline = pipeline_name)
}
gene_sum_tb = bind_rows(gene_sum_tb_ls)
gene_sum_tb = gene_sum_tb %>%
  pivot_wider(names_from = pipeline, values_from = value)
```

```{r}
# Plot comparison of marker genes for cell type across pipelines
gene_sum_tb %>%
  ggplot(aes(x = transcripts, y = preandmrna)) + 
  geom_point(size = 0.1, alpha = 0.5) +
  geom_abline(intercept = 0, slope = 1) +
  facet_wrap(~ name, scales = "free") +
  labs(x = "num_reads_transcripts",
       y = "num_reads_preandmrna") +
  theme_bw()

# Boxplot
gene_sum_tb_2 = gene_sum_tb %>%
  pivot_longer(-c("cell_barcode", "name"), names_to = "pipeline", values_to = "counts")
summ <- gene_sum_tb_2 %>% 
  group_by(pipeline) %>% 
  summarize(median = median(counts))
gene_sum_tb_2 %>%
  ggplot(aes(x = pipeline, y = counts)) +
  geom_boxplot() +
  geom_label(data = summ, aes(x = pipeline, y = median, 
                              label = round(median, 0))) +
  theme_bw() 
```

```{r}
# Plot ratio of astrocyte:qNSC marker genes for astrocyte cells between pipelines
astrocyte_mg = unique(unlist(all.markers$Astrocytes))
astrocyte_genes = t2g %>% filter(gene_name %in% astrocyte_mg) %>% pull(gene_id)
qnsc_mg = unique(unlist(all.markers$qNSCs))
qnsc_genes = t2g %>% filter(gene_name %in% qnsc_mg) %>% pull(gene_id)

gene_type_name = "Astrocyte qNSC ratio"
gene_sum_tb_ls = list()
for(i in seq_along(sce_ls)){
  pipeline_name = names(sce_ls)[i]
  sce_i = sce_ls[[i]]
  
  lib_size_tb = tibble(cell_barcode = astrocyte_cells)
  lib_size_gene_type_astrocyte = colSums(counts(sce_i)[intersect(rownames(counts(sce_i)), astrocyte_genes), astrocyte_cells, drop = F])
  lib_size_gene_type_qnsc = colSums(counts(sce_i)[intersect(rownames(counts(sce_i)), qnsc_genes), astrocyte_cells, drop = F])
  lib_size_gene_type_tb = tibble(cell_barcode = names(lib_size_gene_type_astrocyte),
                                 !!gene_type_name := lib_size_gene_type_astrocyte/lib_size_gene_type_qnsc)
  lib_size_tb = lib_size_tb %>% 
    left_join(., lib_size_gene_type_tb, by = "cell_barcode")
  gene_sum_tb_ls[[pipeline_name]] = lib_size_tb %>%
    pivot_longer(cols = -cell_barcode) %>%
    mutate(pipeline = pipeline_name)
}
gene_sum_tb = bind_rows(gene_sum_tb_ls)
gene_sum_tb = gene_sum_tb %>%
  pivot_wider(names_from = pipeline, values_from = value)

# Plot comparison of marker genes for cell type across pipelines
gene_sum_tb %>%
  ggplot(aes(x = transcripts, y = preandmrna)) + 
  geom_point(size = 0.1, alpha = 0.5) +
  geom_abline(intercept = 0, slope = 1) +
  facet_wrap(~ name, scales = "free") +
  labs(x = "num_reads_transcripts",
       y = "num_reads_preandmrna") +
  theme_bw()
```

```{r}
# Boxplot comparing ratios across pipelines
gene_sum_tb_2 = gene_sum_tb %>%
  pivot_longer(-c("cell_barcode", "name"), names_to = "pipeline", values_to = "ratio")
summ <- gene_sum_tb_2 %>% 
  left_join(., cd, by = c("cell_barcode" = "cell")) %>%
  group_by(pipeline) %>% 
  summarize(median = median(ratio))
gene_sum_tb_2 %>%
  left_join(., cd, by = c("cell_barcode" = "cell")) %>%
  ggplot(aes(x = pipeline, y = ratio)) +
  geom_boxplot() +
  geom_label(data = summ, aes(x = pipeline, y = median, 
                              label = round(median, 2))) +
  labs() +
  theme_bw() 
```

```{r}
# Boxplot comparing ratios across singleR_labels under a given pipeline
summ <- gene_sum_tb_2 %>% 
  filter(pipeline == !!pipeline) %>%
  left_join(., cd, by = c("cell_barcode" = "cell")) %>%
  group_by(singleR_labels_pruned) %>% 
  summarize(median = median(ratio))
gene_sum_tb_2 %>%
  filter(pipeline == !!pipeline) %>%
  left_join(., cd, by = c("cell_barcode" = "cell")) %>%
  ggplot(aes(x = singleR_labels_pruned, y = ratio)) +
  geom_boxplot() +
  geom_label(data = summ, aes(x = singleR_labels_pruned, y = median, 
                              label = round(median, 2))) +
  labs(title = pipeline,
       y = "Astrocyte:qNSC ratio") +
  theme_bw() 
```


```{r}
# Find genes that are most different between pipelines
gene_diff_tb_ls = list()
for(i in seq_along(sce_ls)){
  pipeline_name = names(sce_ls)[i]
  sce_i = sce_ls[[i]]
  gene_diff_tb_ls[[i]] = tibble(pipeline = pipeline_name,
                            gene_id = rownames(counts(sce_i)),
                            sum = rowSums(counts(sce_i)))
}

gene_diff_tb = bind_rows(gene_diff_tb_ls) %>%
  pivot_wider(names_from = pipeline, values_from = sum) %>%
  left_join(., t2g %>% select(gene_id, gene_name), by = "gene_id") %>% # Join to gene name
  mutate(diff = (preandmrna - transcripts)/preandmrna) %>%             # Calculate percent increase from transcripts to preandmrna
  arrange(-diff) %>%                                                   # Arrange by difference
  filter(complete.cases(.))                                            # Remove rows with NA (e.g. genes that are not present in all pipelines)


# Markers for all genes
all_marker_genes_tb_ls = list()
for(cell_type in names(all.markers)){
  all_marker_genes_tb_ls[[cell_type]] = tibble(cell_type = cell_type,
                                               gene_name = unique(unlist(all.markers[[cell_type]])))
}
all_marker_genes_tb = bind_rows(all_marker_genes_tb_ls)
gene_diff_plt = gene_diff_tb %>%
  left_join(., all_marker_genes_tb, by = "gene_name") %>%
  filter(complete.cases(.)) %>%     # Remove genes that are not marker genes
  mutate(group = ceiling((1:n())/1000)) # Assign group for every 1000 genes

# Summary statistics (median difference for each group)
gene_diff_plt %>% 
  group_by(group) %>%
  summarize(median_diff = median(diff)) %>%
  ungroup(group) %>%
  ggplot(aes(x = group, y = median_diff)) +
  geom_bar(stat = "identity") +
  scale_x_continuous(breaks = 1:9) +
  theme_bw()
```

```{r fig.height = 10, fig.width = 8}
library(tidytext) # for reorder_within
# Plot distribution of marker genes for each group 
gene_diff_plt %>%
  group_by(cell_type, group) %>%
  summarize(count = n()) %>%
  ungroup() %>%
  ggplot(aes(x = reorder_within(cell_type, -count, group), y = count)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ group, scales = "free_x", labeller = label_both) +
  labs(x = "Cell type marker genes",
       title = "Distribution of marker genes grouped by difference between pipelines") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))


gene_diff_plt %>%
  group_by(cell_type, group) %>%
  summarize(count = n()) %>%
  ungroup() %>%
  ggplot(aes(x = group, y = count)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ cell_type, scales = "free_x") +
  scale_x_continuous(breaks = 1:9) +
  labs(x = "Cell type marker genes",
       title = "Distribution of marker genes across the groups") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

```{r}
# Plot lib size for different cell types
lib_size_tb = tibble(cell_barcode = colnames(sce_ls[[1]]),
                     ding_labels = colData(sce)$ding_labels)

for(i in seq_along(sce_ls)){
  pipeline_name = names(sce_ls)[i]
  sce_i = sce_ls[[i]]
  
  lib_size = colSums(counts(sce_i))
  lib_size_pipeline_tb = tibble(cell_barcode = colnames(sce_i),
                       !!paste0("lib_size_", pipeline_name) := lib_size)
  lib_size_tb = lib_size_tb %>%
    left_join(., lib_size_pipeline_tb, by = "cell_barcode")
}

lib_size_tb %>%
  ggplot(aes(x = lib_size_transcripts, y = lib_size_preandmrna, color = ding_labels), alpha = 0.5) + 
  facet_wrap(~ ding_labels) + 
  geom_point() +
  geom_abline(slope = 1, intercept = 0) +
  theme_bw()
```

### Distributions

To assess distributions, we want to obtain a homogenous group of cells.

```{r}
source(here("./mouse_cortex/code/distribution-plots-helpers.R"))
```


```{r}
sce = sce_ls[["transcripts"]]

# Subset to cluster
cell_type = "Inhibitory neuron"
sce_sub = sce[, colData(sce)$ding_labels == cell_type & !is.na(colData(sce)$ding_labels)]
dim(sce_sub)
```

#### Within-group clustering

Rerun PCA within cell type subgroup.

```{r}
# Normalize and take log
# clust = quickCluster(sce_sub)
# sce_sub = computeSumFactors(sce_sub, clusters = clust)
# sce_sub = logNormCounts(sce_sub)

set.seed(1)
tic("approx PCA")
sce_sub = runPCA(sce_sub, exprs_values = "logcounts", ntop = ncol(sce), BSPARAM = RandomParam())
toc()
```

```{r}
colData(sce_sub)$cortex_flow_cell = paste(colData(sce_sub)$cortex,
                                         colData(sce_sub)$flow_cell, sep = "_")
colData(sce_sub)$log_lib_size = log(colData(sce_sub)$lib_size)
plotReducedDim(sce_sub, "PCA", colour_by = "log_lib_size", ncomponents = c(1, 2))
```

```{r eval = F}
# K-means clustering
set.seed(1)

# Elbow plot with SSE
sse = c()
for(k in 1:10){
  clust.kmeans <- kmeans(reducedDim(sce_sub, "PCA"), centers = k)
  sse = c(sse, clust.kmeans$tot.withinss)
}
plot(sse)

clust.kmeans <- kmeans(reducedDim(sce_sub, "PCA"), centers = 4)
colData(sce_sub)$labels_kmeans <- factor(clust.kmeans$cluster)
plotReducedDim(sce_sub, "PCA", colour_by = "labels_kmeans", ncomponents = c(1, 2))
```

```{r eval = F}
# Graph-based clustering
g <- buildSNNGraph(sce_sub, k=10, use.dimred = 'PCA')
clust <- igraph::cluster_walktrap(g)$membership

colData(sce_sub)$labels_gb <- factor(clust)
plotReducedDim(sce_sub, "PCA", colour_by="labels_gb")
```

#### Proportion of zero plots

First, I make all the column sums comparable by downsampling (this is a stochastic transformation as opposed to scaling).

```{r}
cortex = "Cortex1"
counts_sub = counts(sce_sub)[, colData(sce_sub)$cortex == cortex]
summary(colSums(counts_sub))

counts_sub_scaled = Down_Sample_Matrix(ceiling(counts_sub))
counts_sub_scaled = counts_sub_scaled[rowSums(counts_sub_scaled) != 0, ]
summary(colSums(counts_sub_scaled))
```

I compute the average expression for every row (x-axis) $\hat{\mu}_i$ and the empirical $P(X_i=0)$, which is the probability that for a given transcript $i$, the count is 0. 

I then compute what $P(X_i=0)$ would be under the model assumptions of binomial or poisson, using parameters estimated from the data. In particular, for a $Binom(n, p_i)$, $n$ is the median number of total counts of cells and $p_i$ is the proportion of counts that were in gene $i$.

Generate and save for all combinations by running `distribution_plots.R`.

```{r}
# Plot P(X_i = 0) against average expression level
prob_out = plot_prob(counts_sub_scaled)
prob_out$plot +
  labs(title = paste(cell_type, cortex, sep = " - "))
```

#### Goodness-of-fit tests

There are 2 issues:

1) Why is quant matrix not all integers?
2) Why are NaNs produced/non-finite finite-difference value errors occur in fitdistr("negative binomial")?

```{r eval = F}
library(MASS)

# Poisson
tic()
p_values_pois = c()
for(i in 1:nrow(counts_sub_scaled)){
  # message(i)
  x = counts_sub_scaled[i,]
  x = ceiling(x) 
  poissonfit = fitdistr(x, "poisson")
  lambda.fit = poissonfit$estimate["lambda"]
  nsamples = length(x)
  f.obs = table(x)
  bins = as.numeric(names(f.obs))
  f.hyp = dpois(bins, lambda = lambda.fit)*nsamples
  chiSquare.pois = sum((f.obs-f.hyp)^2/f.hyp)
  dof = length(bins) - 1
  p_values_pois = c(p_values_pois, 1 - pchisq(chiSquare.pois, df = dof))
}
toc()

# Negative binomial
tic()
p_values_nb = c()
for(i in 1:nrow(counts_sub_scaled)){
  # message(i)
  tryCatch({
  x = counts_sub_scaled[i,]
  x = ceiling(x)
  nbfit = suppressWarnings(fitdistr(x, "negative binomial"))
  r.fit = nbfit$estimate['size']
  p.fit = r.fit / (nbfit$estimate['mu'] + r.fit)
  nsamples = length(x)
  f.obs = table(x)
  bins = as.numeric(names(f.obs))
  f.hyp = dnbinom(bins, size = r.fit, prob = p.fit)*nsamples
  chiSquare.nb = sum((f.obs-f.hyp)^2/f.hyp)
  dof = length(bins) - 1
  p_values_nb = c(p_values_nb, 1 - pchisq(chiSquare.nb, df = dof))}, error = function(e){})
}
toc()

# Put p-values in tibble
print(length(p_values_pois) - length(p_values_nb))
p_value_tb = tibble(distribution = c(rep("Poisson", length(p_values_pois)), rep("NB", length(p_values_nb))),
       p_value = c(p_values_pois, p_values_nb),
       gene = rep(rownames(counts_sub_scaled), 2))

# Histogram of p-values
p_value_tb %>%
  ggplot(aes(x = p_value)) + 
  geom_histogram() +
  facet_wrap(~ distribution) +
  labs(title = cell_type) +
  theme_bw()

# Scatterplot of p-values
emp_probs_0 = apply(counts_sub_scaled, MARGIN = 1, function(r) sum(r==0)/ncol(counts_sub_scaled))
p_value_tb %>%
  pivot_wider(id_cols = gene,
              names_from = "distribution",
              values_from = "p_value") %>%
  mutate(emp_probs_0 = emp_probs_0) %>%
  ggplot(aes(x = Poisson, y = NB)) + 
  geom_point(aes(color = emp_probs_0), alpha = 0.5) +
  labs(x = "poisson_pvalue",
       y = "nb_pvalue") +
  theme_bw()
```


#### BIC log-likelihood

Source: https://github.com/willtownes/scrna2019/blob/028363f04139a58b19143f3058ca8fa4a3533b63/util/functions.R


```{r eval = F}
m = counts_sub_scaled
m = ceiling(m)
summary(colSums(m))
summary(rowSums(m))
m = m[rowSums(m) != 0, ]
dim(m)

bic_tb = tibble(multinomial = mult_bic(m),     # 5 sec
                # dmn = dmn_bic(m),     # Unknown time, need to run as script on cluster
                poisson = poi_bic(m),       # 5 sec
                negative_binomial_1 = nb_bic_1(m),     # 5 sec
                negative_binomial_2 = nb_bic_2(m),   # 15 min
                cell_type_name = cell_type)

# Plot
bic_tb %>% 
  pivot_longer(cols = -cell_type_name) %>%
  ggplot(aes(x = reorder(name, value), y = value)) +
  geom_point(size = 3) +
  labs(title = pipeline,
       x = "Distribution",
       y = "BIC") +
  theme_bw()
```



#### Mean-variance plots

This section has been moved to `distribution-plots.R`.


```{r}

```
